<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>microbiome biomarker analysis toolkit • microbiomeMarker</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="microbiome biomarker analysis toolkit">
<meta property="og:description" content="To date, a number of methods have been developed for microbiome 
    marker discovery based on metagenomic profiles, e.g. LEfSe. However, all of 
    these methods have its own advantages and disadvantages, and none of them is 
    considered standard or universal. Moreover, different programs or softwares 
    may be development using different programming languages, even in different 
    operating systems. Here, we have developed an all-in-one R package 
    microbiomeMarker that integrates commonly used differential analysis 
    methods as well as three machine learning-based approaches, including 
    Logistic regression, Random forest, and Support vector machine, to 
    facilitate the identification of microbiome markers.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">microbiomeMarker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/microbiomeMarker-vignette.html">Tools for microbiome marker identification</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/yiluheihei/microbiomeMarker/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="microbiomemarker">microbiomeMarker<a class="anchor" aria-label="anchor" href="#microbiomemarker"></a>
</h1></div>
<p><a href="https://github.com/yiluheihei/microbiomeMarker" class="external-link"></a><img src="reference/figures/microbiomeMarker.png" height="150" align="right" style="float:right; height:150px;"></p>
<!-- badges: start -->
<p><strong><em>microbiomeMarker</em></strong> is still under development, your suggestion and contribution will be highly appreciated. If you think this project is helpful to you, you can give this project a <span class="emoji" data-emoji="star">⭐</span>.</p>
<div class="section level2">
<h2 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h2>
<p>The aim of this package is to build a unified toolbox in R for microbiome biomarker discovery by integrating existing widely used differential analysis methods.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the package from Bioconductor directly:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"microbiomeMarker"</span><span class="op">)</span></span></code></pre></div>
<p>Or install the development version of the package from <a href="https://github.com/yiluheihei/microbiomeMarker" class="external-link">Github</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"yiluheihei/microbiomeMarker"</span><span class="op">)</span></span></code></pre></div>
<p>For more details on how to use <strong><em>microbiomeMarker</em></strong>, please see the help page or <a href="https://yiluheihei.github.io/microbiomeMarker/index.html" class="external-link">website</a> of our package.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Kindly cite as follows:</p>
<p><strong>Yang Cao</strong>, Qingyang Dong, Dan Wang, Pengcheng Zhang, Ying Liu, Chao Niu, microbiomeMarker: an R/Bioconductor package for microbiome marker identification and visualization. Bioinformatics, 2022, btac438. doi: <a href="https://doi.org/10.1093/bioinformatics/btac438" class="external-link">10.1093/bioinformatics/btac438</a></p>
<p>Please cite the corresponding methods paper too:</p>
<ul>
<li>LEfSe: Segata, Nicola, Jacques Izard, et al. 2011. Metagenomic Biomarker Discovery and Explanation. Genome Biology 12 (6): 1–18. doi: <a href="https://doi.org/10.1186/gb-2011-12-6-r60" class="external-link">10.1186/gb-2011-12-6-r60</a>
</li>
<li>metagenomeSeq: Paulson, Joseph N, O Colin Stine, et al. 2013. Differential Abundance Analysis for Microbial Marker-Gene Surveys. Nature Methods 10 (12): 1200–1202. doi: <a href="https://doi.org/10.1038/nmeth.2658" class="external-link">10.1038/nmeth.2658</a>
</li>
<li>ANCOM: Mandal, Siddhartha, Will Van Treuren, et al. 2015. Analysis of Composition of Microbiomes: A Novel Method for Studying Microbial Composition. Microbial Ecology in Health and Disease 26 (1): 27663. doi: <a href="https://doi.org/10.3402/mehd.v26.27663" class="external-link">10.3402/mehd.v26.27663</a>
</li>
<li>ANCOMBC: Lin, Huang, and Shyamal Das Peddada. 2020. Analysis of Compositions of Microbiomes with Bias Correction. Nature Communications 11 (1): 1–11. doi: <a href="https://doi.org/10.1038/s41522-020-00160-w" class="external-link">10.1038/s41522-020-00160-w</a>
</li>
<li>ALDEx2: Fernandes, Andrew D, Jennifer Ns Reid, et al. 2014. Unifying the Analysis of High-Throughput Sequencing Datasets: Characterizing Rna-Seq, 16S rRNA Gene Sequencing and Selective Growth Experiments by Compositional Data Analysis. Genome Biology 15(2): 1–17. doi: <a href="https://doi.org/10.1186/2049-2618-2-15" class="external-link">10.1186/2049-2618-2-15</a>
</li>
<li>edgeR: Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. EdgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data. Bioinformatics 26 (1): 139–40. doi: <a href="https://doi.org/10.1093/bioinformatics/btp616" class="external-link">10.1093/bioinformatics/btp616</a>
</li>
<li>DESeq2: Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. Moderated Estimation of Fold Change and Dispersion for Rna-Seq Data with Deseq2. Genome Biology 15 (12): 1–21. doi: <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">10.1186/s13059-014-0550-8</a>
</li>
<li>limma-voom: Law, Charity W, Yunshun Chen, et al. 2014. Voom: Precision Weights Unlock Linear Model Analysis Tools for Rna-Seq Read Counts. Genome biology, 15(2), 1-17. doi: <a href="https://doi.org/10.1186/gb-2014-15-2-r29" class="external-link">10.1186/gb-2014-15-2-r29</a>
</li>
</ul>
<div class="section level3">
<h3 id="publications-citing-microbiomemarker">Publications citing microbiomeMarker<a class="anchor" aria-label="anchor" href="#publications-citing-microbiomemarker"></a>
</h3>
<ul>
<li>Jorge F, Froissard C, Dheilly N M, et al. Bacterial community dynamics following antibiotic exposure in a trematode parasite. International journal for parasitology, 2022, 52(5): 265-274. <a href="https://doi.org/10.1016/j.ijpara.2021.11.006" class="external-link uri">https://doi.org/10.1016/j.ijpara.2021.11.006</a>
</li>
<li>Ghosh A, Thakur M, Sharma L K, et al. Linking gut microbiome with the feeding behavior of the Arunachal macaque (Macaca munzala). Scientific reports, 2021, 11(1): 1-10. <a href="https://doi.org/10.1038/s41598-021-01316-0" class="external-link uri">https://doi.org/10.1038/s41598-021-01316-0</a>
</li>
<li>Doi R, Wu Y, Kawai Y, et al. Transition and regulation mechanism of bacteria biota in Kishu saba-narezushi (mackerel narezushi) during its fermentation step. Journal of bioscience and bioengineering, 2021, 132(6): 606-612. <a href="https://doi.org/10.1016/j.jbiosc.2021.09.002" class="external-link uri">https://doi.org/10.1016/j.jbiosc.2021.09.002</a>
</li>
<li>Nielsen K R, Ingham A C, Midjord J, et al. Similar Gut Bacterial Composition Between Patients With Ulcerative Colitis and Healthy Controls in a High Incidence Population: A Cross-sectional Study of the Faroe Islands IBD Cohort. Inflammatory Bowel Diseases. <a href="https://doi.org/10.1093/ibd/izab355" class="external-link uri">https://doi.org/10.1093/ibd/izab355</a>
</li>
<li>Prevel R, Enaud R, Orieux A, et al. Gut bacteriobiota and mycobiota are both associated with Day-28 mortality among critically ill patients. Critical Care, 2022, 26(1): 1-9. <a href="https://doi.org/10.1186/s13054-022-03980-8" class="external-link uri">https://doi.org/10.1186/s13054-022-03980-8</a>
</li>
<li>Tandon K, Chiou Y J, Yu S P, et al. Microbiome restructuring: dominant coral bacterium Endozoicomonas species display differential adaptive capabilities to environmental changes[J]. bioRxiv, 2021. <a href="https://doi.org/10.1101/2021.10.31.466697" class="external-link uri">https://doi.org/10.1101/2021.10.31.466697</a>
</li>
<li>Dai D, Zhu J, Sun C, et al. GMrepo v2: a curated human gut microbiome database with special focus on disease markers and cross-dataset comparison. Nucleic acids research, 2022, 50(D1): D777-D784. <a href="https://doi.org/10.1093/nar/gkab1019" class="external-link uri">https://doi.org/10.1093/nar/gkab1019</a>
</li>
<li>Molinero N, Taladrid D, Zorraquín-Peña I, et al. Ulcerative Colitis Seems to Imply Oral Microbiome Dysbiosis. Current Issues in Molecular Biology, 2022, 44(4): 1513-1527. <a href="https://doi.org/10.3390/cimb44040103" class="external-link uri">https://doi.org/10.3390/cimb44040103</a>
</li>
<li>Ricci F, Tandon K, Black J R, et al. Host Traits and Phylogeny Contribute to Shaping Coral-Bacterial Symbioses[J]. Msystems, 2022, 7(2): e00044-22. <a href="https://doi.org/10.1128/msystems.00044-22" class="external-link uri">https://doi.org/10.1128/msystems.00044-22</a>
</li>
<li>Chavarria K A, Saltonstall K, Vinda J, et al. Land use influences stream bacterial communities in lowland tropical watersheds. Scientific reports, 2021, 11(1): 1-12. <a href="https://doi.org/10.1038/s41598-021-01193-7" class="external-link uri">https://doi.org/10.1038/s41598-021-01193-7</a>
</li>
<li>Lu H, Gao N L, Tong F, et al. Alterations of the Human Lung and Gut Microbiomes in Non-Small Cell Lung Carcinomas and Distant Metastasis. Microbiology spectrum, 2021, 9(3): e00802-21. <a href="https://doi.org/10.1128/Spectrum.00802-21" class="external-link uri">https://doi.org/10.1128/Spectrum.00802-21</a>
</li>
<li>Ingham A C, Kielsen K, Mordhorst H, et al. Microbiota long-term dynamics and prediction of acute graft-versus-host disease in pediatric allogeneic stem cell transplantation[J]. Microbiome, 2021, 9(1): 1-28. <a href="https://doi.org/10.1186/s40168-021-01100-2" class="external-link uri">https://doi.org/10.1186/s40168-021-01100-2</a>
</li>
<li>Wang R, Cao S, Bashir M E H, et al. Microbial metabolite butyrate-prodrug polymeric micelles promote gut health and treat food allergies. bioRxiv,<ol start="2022" style="list-style-type: decimal"><li><a href="https://doi.org/10.1101/2022.05.01.490224" class="external-link uri">https://doi.org/10.1101/2022.05.01.490224</a></li></ol>
</li>
<li>Shanmugam G, Lee SH, Jeon J. EzMAP: Easy Microbiome Analysis Platform. BMC bioinformatics. 2021 Dec;22(1):1-0. <a href="https://doi.org/10.1186/s12859-021-04106-7" class="external-link uri">https://doi.org/10.1186/s12859-021-04106-7</a>
</li>
<li>Altaib H, Nakamura K, Abe M, et al. Differences in the concentration of the fecal neurotransmitters GABA and glutamate are associated with microbial composition among healthy human subjects. Microorganisms. 2021. Feb;9(2):378. <a href="https://doi.org/10.3390/microorganisms9020378" class="external-link uri">https://doi.org/10.3390/microorganisms9020378</a>
</li>
<li>Künstner A, Aherrahrou R, Hirose M, et al. Effect of Differences in the Microbiome of Cyp17a1-Deficient Mice on Atherosclerotic Background. Cells. 2021 Jun;10(6):1292. <a href="https://doi.org/10.3390/cells10061292" class="external-link uri">https://doi.org/10.3390/cells10061292</a>
</li>
<li>Ingham AC, Urth TR, Sieber RN, et al. Dynamics of the human nasal microbiota and Staphylococcus aureus CC398 carriage in pig truck drivers across one workweek. Applied and Environmental Microbiology. 2021 Jun 30:AEM-01225. <a href="https://doi.org/10.1128/AEM.01225-21" class="external-link uri">https://doi.org/10.1128/AEM.01225-21</a>
</li>
<li>Shibata T, Nakagawa M, Coleman HN, et al. Evaluation of DNA extraction protocols from liquid-based cytology specimens for studying cervical microbiota. Plos one 16, no. 8 2021. <a href="https://doi.org/10.1371/journal.pone.0237556" class="external-link uri">https://doi.org/10.1371/journal.pone.0237556</a>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="question">Question<a class="anchor" aria-label="anchor" href="#question"></a>
</h2>
<p>If you have any question, please file an issue on the issue tracker following the instructions in the issue template:</p>
<p>Please briefly describe your problem, what output actually happened, and what output you expect.</p>
<p>Please provide a minimal reproducible example. For more details on how to make a great minimal reproducible example, see <a href="https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example" class="external-link uri">https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example</a> and <a href="https://www.tidyverse.org/help/#reprex" class="external-link uri">https://www.tidyverse.org/help/#reprex</a>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>Brief description of the problem</span>
<span id="cb3-2"><a href="#cb3-2"></a></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co"># insert minimal reprducible example here</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="acknowledgement">Acknowledgement<a class="anchor" aria-label="anchor" href="#acknowledgement"></a>
</h2>
<p>We thanks all the developers of the methods integrated into our package.</p>
<ul>
<li>
<a href="https://bitbucket.org/biobakery/biobakery/wiki/lefse" class="external-link">lefse python script</a>, The main lefse code are translated from <strong><em>lefse python script</em></strong>,</li>
<li>
<a href="https://github.com/lch14forever/microbiomeViz" class="external-link">microbiomeViz</a>, cladogram visualization of lefse is modified from <strong><em>microbiomeViz</em></strong>.</li>
<li>
<a href="https://github.com/joey711/phyloseq" class="external-link">phyloseq</a>, the main data structures used in <strong><em>microbiomeMarker</em></strong> are from or inherit from <code>phyloseq-class</code> in package <strong><em>phyloseq</em></strong>.</li>
<li>
<a href="https://github.com/YuLab-SMU/MicrobiotaProcess" class="external-link">MicrobiotaProcess</a>, function <code><a href="reference/import_dada2.html">import_dada2()</a></code> and <code><a href="reference/import_qiime2.html">import_qiime2()</a></code> are modified from the <code>MicrobiotaProcess::import_dada2()</code>.</li>
<li>
<a href="https://github.com/jbisanz/qiime2R" class="external-link">qiime2R</a>, <code><a href="reference/import_qiime2.html">import_qiime2()</a></code> are refer to the functions in <strong><em>qiime2R</em></strong>.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/microbiomeMarker" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/yiluheihei/microbiomeMarker/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/yiluheihei/microbiomeMarker/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing microbiomeMarker</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yang Cao <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/microbiomeMarker" class="external-link"><img src="https://img.shields.io/badge/release%20version-1.2.1-green.svg"></a></li>
<li><a href="https://github.com/yiluheihei/microbiomeMarker" class="external-link"><img src="https://img.shields.io/badge/devel%20version-1.3.2-green.svg"></a></li>
<li><a href="https://www.bioconductor.org/packages/devel/bioc/html/microbiomeMarker.html#archives" class="external-link"><img src="http://www.bioconductor.org/shields/availability/devel/microbiomeMarker.svg" alt="platform"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/microbiomeMarker" class="external-link"><img src="http://www.bioconductor.org/shields/build/release/bioc/microbiomeMarker.svg" alt="BioC status"></a></li>
<li><a href="https://www.bioconductor.org/packages/devel/bioc/html/microbiomeMarker.html#since" class="external-link"><img src="http://www.bioconductor.org/shields/years-in-bioc/microbiomeMarker.svg" alt="Bioc years"></a></li>
<li><a href="https://github.com/yiluheihei/microbiomeMarker/actions" class="external-link"><img src="https://github.com/yiluheihei/microbiomeMarker/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
<li><a href="https://github.com/yiluheihei/microbiomeMarker/blob/master/LICENSE.md" class="external-link"><img src="https://img.shields.io/badge/License-GPLv3-blue.svg" alt="License: GPL v3"></a></li>
<li><a href="https://codecov.io/gh/yiluheihei/microbiomeMarker?branch=master" class="external-link"><img src="https://codecov.io/gh/yiluheihei/microbiomeMarker/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://zenodo.org/badge/latestdoi/215731961" class="external-link"><img src="https://zenodo.org/badge/215731961.svg" alt="DOI"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yang Cao.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
