<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>microbiome biomarker analysis • microbiomeMarker</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="microbiome biomarker analysis">
<meta property="og:description" content="microbiome biomarker analysis.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">microbiomeMarker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/yiluheihei/microbiomeMarker/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="microbiomemarker" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#microbiomemarker" class="anchor"></a>microbiomeMarker</h1></div>
<p><a href="https://github.com/yiluheihei/microbiomeMarker"></a><img src="reference/figures/microbiomeMarker.png" height="150" align="right"></p>
<!-- badges: start -->
<p><a href="https://github.com/yiluheihei/microbiomeMarker/actions"><img src="https://github.com/yiluheihei/microbiomeMarker/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a> <a href="https://github.com/yiluheihei/microbiomeMarker/blob/master/LICENSE.md"><img src="https://img.shields.io/badge/License-GPLv3-blue.svg" alt="License: GPL v3"></a> <a href="https://codecov.io/gh/yiluheihei/microbiomeMarker?branch=master"><img src="https://codecov.io/gh/yiluheihei/microbiomeMarker/branch/master/graph/badge.svg" alt="Codecov test coverage"></a> <a href="https://zenodo.org/badge/latestdoi/215731961"><img src="https://zenodo.org/badge/215731961.svg" alt="DOI"></a> <!-- badges: end --></p>
<p><strong>If you think this project is helpful to you, you can give this project a star</strong> <span class="emoji" data-emoji="star">⭐</span></p>
<div id="motivation" class="section level2">
<h2 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h2>
<p><strong>The aim of this package is to build a unified toolbox in R for mcirobiome biomarker discovery by integrating various existing methods.</strong></p>
<p>Many statistical methods have been proposed to discovery the microbiome biomaker by compare the taxon abundance between different classes. Some methods developed specifically for microbial community, such as linear discriminant analysis (LDA) effect size (LEfSe) (Segata et al. 2011), metagenomeSeq (Paulson et al. 2013); and some methods developed specifically for RNA-Seq data, such as DESeq2 (Love, Huber, and Anders 2014) and edgeR (Robinson, McCarthy, and Smyth 2009), have been proposed for microbiome biomarker discovery. We usually use several methods for microbiome biomarker discovery and compare the results, which requires multiple tools developed in different programming, even in different OS.</p>
<p><strong>microbiomeMarker</strong> take the <code>phyloseq-class</code> object in package <a href="https://github.com/joey711/phyloseq">phyloseq</a> as input, since <strong>phyloseq</strong> is the most popular R package in microbiome analysis and with phyloseq you can easily import taxon abundance and phylogenetic tree of taxon output from common microbiome bioinformatics platforms, such as <a href="https://benjjneb.github.io/dada2/">DADA2</a> and <a href="https://qiime2.org/">qiime2</a>.</p>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>Kindly cite as follows: Yang Cao (2020). microbiomeMarker: microbiome biomarker analysis. R package version 0.0.1.9000. <a href="https://github.com/yiluheihei/microbiomeMarker" class="uri">https://github.com/yiluheihei/microbiomeMarker</a>. DOI: <a href="https://doi.org/10.5281/zenodo.3749415">10.5281/zenodo.3749415</a>.</p>
<div id="publications-citing-microbiomemarker" class="section level3">
<h3 class="hasAnchor">
<a href="#publications-citing-microbiomemarker" class="anchor"></a>Publications citing microbiomeMarker</h3>
<ul>
<li><a href="https://doi.org/10.1186/s12859-021-04106-7">Shanmugam G, Lee SH, Jeon J. EzMAP: Easy Microbiome Analysis Platform. BMC bioinformatics. 2021 Dec;22(1):1-0</a></li>
<li><a href="https://doi.org/10.3390/microorganisms9020378">Altaib H, Nakamura K, Abe M, Badr Y, Yanase E, Nomura I, Suzuki T. Differences in the concentration of the fecal neurotransmitters GABA and glutamate are associated with microbial composition among healthy human subjects. Microorganisms. 2021 Feb;9(2):378</a></li>
<li><a href="https://doi.org/10.1101/2021.02.19.21252040">Ingham AC, Kielsen K, Mordhorst H, Ifversen M, Aarestrup FM, Müller KG, Pamp SJ. Microbiota long-term dynamics and prediction of acute graft-versus-host-disease in pediatric allogeneic stem cell transplantation. medRxiv. 2021 Jan 1</a></li>
<li><a href="https://doi.org/10.3390/cells10061292">Künstner A, Aherrahrou R, Hirose M, Bruse P, Ibrahim SM, Busch H, Erdmann J, Aherrahrou Z. Effect of Differences in the Microbiome of Cyp17a1-Deficient Mice on Atherosclerotic Background. Cells. 2021 Jun;10(6):1292</a></li>
</ul>
</div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the package directly from github</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org">remotes</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"yiluheihei/microbiomeMarker"</span><span class="op">)</span></code></pre></div>
</div>
<div id="data-import" class="section level2">
<h2 class="hasAnchor">
<a href="#data-import" class="anchor"></a>Data import</h2>
<p>Since <a href="https://github.com/joey711/phyloseq">phyloseq</a> objects are a great data-standard for microbiome data in R, the core functions in <strong>microbiomeMarker</strong> take <code>phylosq</code> object as input. Conveniently, <strong>microbiomeMarker</strong> provides features to import external data files form two common tools of microbiome analysis, <a href="http://qiime.org/">qiime2</a> and <a href="https://benjjneb.github.io/dada2">dada2</a>.</p>
<div id="import-from-dada2" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-dada2" class="anchor"></a>Import from dada2</h3>
<p>The output of the <a href="https://benjjneb.github.io/dada2">dada2</a> pipeline is a feature table of amplicon sequence variants (an ASV table): A matrix with rows corresponding to samples and columns to ASVs, in which the value of each entry is the number of times that ASV was observed in that sample. This table is analogous to the traditional OTU table. Conveniently, taxa names are saved as</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yiluheihei/microbiomeMarker">microbiomeMarker</a></span><span class="op">)</span>

<span class="va">seq_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dada2_seqtab.rds"</span>,
  package<span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span><span class="op">)</span>
<span class="va">tax_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dada2_taxtab.rds"</span>,
 package<span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sam_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dada2_samdata.txt"</span>,
 package<span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_dada2.html">import_dada2</a></span><span class="op">(</span>seq_tab <span class="op">=</span> <span class="va">seq_tab</span>, tax_tab <span class="op">=</span> <span class="va">tax_tab</span>, sam_tab <span class="op">=</span> <span class="va">sam_tab</span><span class="op">)</span>
<span class="va">ps</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 232 taxa and 20 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 20 samples by 4 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 232 taxa by 6 taxonomic ranks ]</span>
<span class="co">#&gt; refseq()      DNAStringSet:      [ 232 reference sequences ]</span></code></pre></div>
</div>
<div id="import-from-qiime2" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-qiime2" class="anchor"></a>Import from qiime2</h3>
<p><a href="http://qiime.org/">qiime2</a> is the most widely used software for metagenomic analysis. User can import the feature table, taxonomic table, phylogenetic tree, representative sequence and sample metadata from qiime2 using <code><a href="reference/import_qiime2.html">import_qiime2()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">otuqza_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"table.qza"</span>,package <span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span>
<span class="va">taxaqza_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"taxonomy.qza"</span>,package <span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span>
<span class="va">sample_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata"</span>, <span class="st">"sample-metadata.tsv"</span>,
  package <span class="op">=</span> <span class="st">"microbiomeMarker"</span>
<span class="op">)</span>
<span class="va">treeqza_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tree.qza"</span>,package <span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_qiime2.html">import_qiime2</a></span><span class="op">(</span>
  otu_qza <span class="op">=</span> <span class="va">otuqza_file</span>, taxa_qza <span class="op">=</span> <span class="va">taxaqza_file</span>,
  sam_tab <span class="op">=</span> <span class="va">sample_file</span>, tree_qza <span class="op">=</span> <span class="va">treeqza_file</span>
<span class="op">)</span>
<span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="co">#&gt; Also defined by 'tidytree'</span>
<span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="co">#&gt; Also defined by 'tidytree'</span>
<span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="co">#&gt; Also defined by 'tidytree'</span>
<span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="co">#&gt; Also defined by 'tidytree'</span>
<span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="co">#&gt; Also defined by 'tidytree'</span>
<span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="co">#&gt; Also defined by 'tidytree'</span>
<span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="co">#&gt; Also defined by 'tidytree'</span>
<span class="va">ps</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 770 taxa and 34 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 34 samples by 9 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 770 taxa by 7 taxonomic ranks ]</span>
<span class="co">#&gt; phy_tree()    Phylogenetic Tree: [ 770 tips and 768 internal nodes ]</span></code></pre></div>
</div>
<div id="import-from-tab-delimited-input-file-of-biobakery-lefse" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-tab-delimited-input-file-of-biobakery-lefse" class="anchor"></a>Import from tab-delimited input file of biobakery lefse</h3>
<p>For <a href="https://huttenhower.sph.harvard.edu/lefse/">biobakey lefse</a> (a <a href="http://huttenhower.sph.harvard.edu/galaxy">Galaxy module</a>, a Conda formula, a Docker image, and included in bioBakery (VM and cloud).), the input file must be a tab-delimited text, consists of a list of numerical features, the class vector and optionally the subclass and subject vectors. The features can be read counts directly or abundance floating-point values more generally, and the first field is the name of the feature. Class, subclass and subject vectors have a name (the first field) and a list of non-numerical strings. <a href="https://huttenhower.sph.harvard.edu/lefse/">biobakery lefse</a>. User can import the input file suitable for <a href="https://huttenhower.sph.harvard.edu/lefse/">biobakery lefse</a> to <code>phyloseq</code> object using <code><a href="reference/import_biobakery_lefse_in.html">import_biobakery_lefse_in()</a></code></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata"</span>,
  <span class="st">"hmp_small_aerobiosis.txt"</span>,
  package <span class="op">=</span> <span class="st">"microbiomeMarker"</span>
<span class="op">)</span>
<span class="co"># six level of taxonomic ranks,</span>
<span class="co"># meta data: row 1 represents class (oxygen_availability),</span>
<span class="co"># row 2 represents subclass (body_site), row 3 represents subject (subject_id)</span>
<span class="va">hmp_oxygen</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_biobakery_lefse_in.html">import_biobakery_lefse_in</a></span><span class="op">(</span>
  <span class="va">file</span>,
  ranks_prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"k"</span>, <span class="st">"p"</span>, <span class="st">"c"</span>, <span class="st">"o"</span>, <span class="st">"f"</span>, <span class="st">"g"</span><span class="op">)</span>,
  meta_rows <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,
<span class="op">)</span>
<span class="va">hmp_oxygen</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 928 taxa and 55 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 55 samples by 3 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 928 taxa by 1 taxonomic ranks ]</span></code></pre></div>
</div>
<div id="other-import-functions-reexport-from-phyloseq" class="section level3">
<h3 class="hasAnchor">
<a href="#other-import-functions-reexport-from-phyloseq" class="anchor"></a>Other import functions reexport from phyloseq</h3>
<p><strong>microbiomeMarker</strong> reexports three import functions from <strong>phyloseq</strong>, including <code><a href="https://rdrr.io/pkg/phyloseq/man/import_biom.html">import_biom()</a></code>, <code><a href="https://rdrr.io/pkg/phyloseq/man/import_qiime.html">import_qiime()</a></code> and <code><a href="https://rdrr.io/pkg/phyloseq/man/import_mothur.html">import_mothur()</a></code>, to help users to import data from <a href="http://biom-format.org/">biom file</a>, and output from <a href="http://www.qiime.org/">qiime</a> and <a href="http://www.mothur.org/">mothur</a>. More details on these three import functions can be see from <a href="https://joey711.github.io/phyloseq/import-data.html#the_import_family_of_functions">here</a>.</p>
<p>Users can also import the external files into <code>phyloseq</code> object manually. For more details on how to create <code>phyloseq</code> object from manually imported data, please see <a href="http://joey711.github.io/phyloseq/import-data.html#manual">this tutorial</a>.</p>
</div>
</div>
<div id="lefse" class="section level2">
<h2 class="hasAnchor">
<a href="#lefse" class="anchor"></a>LEfSe</h2>
<p>Curently, LEfSe is the most used tool for microbiome biomarker discovery, and the first method to integrate to <strong>microbiomeMarker</strong> is LEfSe.</p>
<div id="lefse-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#lefse-analysis" class="anchor"></a>lefse analysis</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>

<span class="co"># sample data from lefse python script. The dataset contains 30 abundance </span>
<span class="co"># profiles (obtained processing the 16S reads with RDP) belonging to 10 rag2 </span>
<span class="co"># (control) and 20 truc (case) mice</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"spontaneous_colitis"</span><span class="op">)</span>
<span class="co"># add prefix of ranks</span>
<span class="va">mm_lefse</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_lefse.html">run_lefse</a></span><span class="op">(</span>
  <span class="va">kostic_crc</span>,
  wilcoxon_cutoff <span class="op">=</span> <span class="fl">0.01</span>,
  group <span class="op">=</span> <span class="st">"DIAGNOSIS"</span>,
  kw_cutoff <span class="op">=</span> <span class="fl">0.01</span>,
  multigrp_strat <span class="op">=</span> <span class="cn">TRUE</span>,
  lda_cutoff <span class="op">=</span> <span class="fl">4</span>,
<span class="op">)</span>
<span class="co"># lefse return a microbioMarker class inherits from phyloseq</span>
<span class="va">mm_lefse</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ CPM ]</span>
<span class="co">#&gt; microbiome marker identity method: [ lefse ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 20 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 960 taxa and  177 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 177 samples by  71 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 960 taxa by 1 taxonomic ranks ]</span></code></pre></div>
<p>The microbiome biomarker information was stored in a new data structure <code>marker_table-class</code> inherited from <code>data.frame</code>, and you can access it by using <code><a href="reference/marker_table.html">marker_table()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">mm_lefse</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;                                                                                                              feature</span>
<span class="co">#&gt; marker1                                                                      k__Bacteria|p__Firmicutes|c__Clostridia</span>
<span class="co">#&gt; marker2                                                     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales</span>
<span class="co">#&gt; marker3                                                                                    k__Bacteria|p__Firmicutes</span>
<span class="co">#&gt; marker4                               k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae</span>
<span class="co">#&gt; marker5 k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroidaceae_Bacteroides</span>
<span class="co">#&gt; marker6                                  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae</span>
<span class="co">#&gt;         enrich_group   ef_lda       pvalue         padj</span>
<span class="co">#&gt; marker1      Healthy 4.938091 2.577541e-06 2.577541e-06</span>
<span class="co">#&gt; marker2      Healthy 4.938091 2.577541e-06 2.577541e-06</span>
<span class="co">#&gt; marker3      Healthy 4.842391 1.425032e-04 1.425032e-04</span>
<span class="co">#&gt; marker4      Healthy 4.820702 3.706059e-04 3.706059e-04</span>
<span class="co">#&gt; marker5      Healthy 4.820702 3.706059e-04 3.706059e-04</span>
<span class="co">#&gt; marker6      Healthy 4.727401 6.529659e-07 6.529659e-07</span></code></pre></div>
</div>
<div id="visualization-of-the-result-of-lefse-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#visualization-of-the-result-of-lefse-analysis" class="anchor"></a>Visualization of the result of lefse analysis</h3>
<p>Bar plot for output of lefse:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plot_ef_bar.html">plot_ef_bar</a></span><span class="op">(</span><span class="va">mm_lefse</span>, label_level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Healthy"</span> <span class="op">=</span> <span class="st">"blue"</span>, <span class="st">"Tumor"</span> <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-lefse-barplot-1.png"><!-- --></p>
</div>
</div>
<div id="statistical-analysis-stamp" class="section level2">
<h2 class="hasAnchor">
<a href="#statistical-analysis-stamp" class="anchor"></a>statistical analysis (stamp)</h2>
<p>STAMP (Parks et al. 2014) is a widely-used graphical software package that provides “best pratices” in choose appropriate statisticalmethods for microbial taxonomic and functional analysis. Users can tests for both two groups or multiple groups, and effect sizes and confidence intervals are supported that allows critical assessment of the biological relevancy of test results. Here, <strong>microbiomeMarker</strong> also integrates the statistical methods used in STAMP for microbial comparison analysis between two-groups and multiple-groups.</p>
<div id="statitical-analysis-between-two-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#statitical-analysis-between-two-groups" class="anchor"></a>Statitical analysis between two groups</h3>
<p>Function <code><a href="reference/run_test_two_groups.html">run_test_two_groups()</a></code> is developed for statistical test between two groups, and three test methods are provided: welch test, t test and white test.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"enterotypes_arumugam"</span><span class="op">)</span>
<span class="co"># take welch test for example</span>
<span class="va">two_group_welch</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_test_two_groups.html">run_test_two_groups</a></span><span class="op">(</span>
  <span class="va">enterotypes_arumugam</span>, 
  group <span class="op">=</span> <span class="st">"Gender"</span>, 
  method <span class="op">=</span> <span class="st">"welch.test"</span>
<span class="op">)</span>

<span class="co"># three significantly differential genera (marker)</span>
<span class="va">two_group_welch</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ TSS ]</span>
<span class="co">#&gt; microbiome marker identity method: [ welch.test ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 3 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 244 taxa and  39 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 39 samples by  9 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 244 taxa by 1 taxonomic ranks ]</span>
<span class="co"># details of result of the three markers</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">two_group_welch</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;                                     feature enrich_group  ef_diff_mean</span>
<span class="co">#&gt; marker1     p__Firmicutes|g__Heliobacterium            M -8.542172e-06</span>
<span class="co">#&gt; marker2         p__Firmicutes|g__Parvimonas            M -1.339857e-05</span>
<span class="co">#&gt; marker3 p__Firmicutes|g__Peptostreptococcus            M -6.695045e-05</span>
<span class="co">#&gt;             pvalue       padj</span>
<span class="co">#&gt; marker1 0.02940341 0.02940341</span>
<span class="co">#&gt; marker2 0.03281399 0.03281399</span>
<span class="co">#&gt; marker3 0.01714937 0.01714937</span></code></pre></div>
</div>
<div id="statistical-analysis-multiple-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#statistical-analysis-multiple-groups" class="anchor"></a>Statistical analysis multiple groups</h3>
<p>Function <code><a href="reference/run_test_multiple_groups.html">run_test_multiple_groups()</a></code> is constructed for statistical test for multiple groups, two test method are provided: anova and kruskal test.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># three groups</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html">subset_samples</a></span><span class="op">(</span>
  <span class="va">enterotypes_arumugam</span>,
  <span class="va">Enterotype</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Enterotype 3"</span>, <span class="st">"Enterotype 2"</span>, <span class="st">"Enterotype 1"</span><span class="op">)</span>
<span class="op">)</span>

<span class="va">multiple_group_anova</span> <span class="op">&lt;-</span>  <span class="fu"><a href="reference/run_test_multiple_groups.html">run_test_multiple_groups</a></span><span class="op">(</span>
  <span class="va">ps</span>,
  group <span class="op">=</span> <span class="st">"Enterotype"</span>, 
  method <span class="op">=</span> <span class="st">"anova"</span>
<span class="op">)</span>

<span class="co"># 24 markers</span>
<span class="va">multiple_group_anova</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ TSS ]</span>
<span class="co">#&gt; microbiome marker identity method: [ anova ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 24 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 238 taxa and  32 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 32 samples by  9 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 238 taxa by 1 taxonomic ranks ]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">multiple_group_anova</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;                                     feature enrich_group ef_eta_squared</span>
<span class="co">#&gt; marker1                    p__Bacteroidetes Enterotype 1      0.5821619</span>
<span class="co">#&gt; marker2                     p__Unclassified Enterotype 3      0.4497271</span>
<span class="co">#&gt; marker3      p__Actinobacteria|g__Scardovia Enterotype 2      0.2196652</span>
<span class="co">#&gt; marker4       p__Bacteroidetes|g__Alistipes Enterotype 3      0.2001541</span>
<span class="co">#&gt; marker5     p__Bacteroidetes|g__Bacteroides Enterotype 1      0.7633661</span>
<span class="co">#&gt; marker6 p__Bacteroidetes|g__Parabacteroides Enterotype 1      0.2582573</span>
<span class="co">#&gt;               pvalue         padj</span>
<span class="co">#&gt; marker1 3.196070e-06 3.196070e-06</span>
<span class="co">#&gt; marker2 1.731342e-04 1.731342e-04</span>
<span class="co">#&gt; marker3 2.742042e-02 2.742042e-02</span>
<span class="co">#&gt; marker4 3.922758e-02 3.922758e-02</span>
<span class="co">#&gt; marker5 8.396825e-10 8.396825e-10</span>
<span class="co">#&gt; marker6 1.314233e-02 1.314233e-02</span></code></pre></div>
<p>The result of multiple group statistic specified whether the means of all groups is equal or not. To identify which pairs of groups may differ from each other, post-hoc test must be performed.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pht</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_posthoc_test.html">run_posthoc_test</a></span><span class="op">(</span><span class="va">ps</span>, group <span class="op">=</span> <span class="st">"Enterotype"</span><span class="op">)</span>
<span class="va">pht</span>
<span class="co">#&gt; postHocTest-class object</span>
<span class="co">#&gt; Pairwise test result of 238  features,  DataFrameList object, each DataFrame has five variables:</span>
<span class="co">#&gt;         comparions    : pair groups to test which separated by '-'</span>
<span class="co">#&gt;         diff_mean: difference in mean proportions</span>
<span class="co">#&gt;         pvalue        : post hoc test p values</span>
<span class="co">#&gt;         ci_lower : lower confidence interval</span>
<span class="co">#&gt;         ci_upper : upper confidence interval</span>
<span class="co">#&gt; Posthoc multiple comparisons of means  using  tukey  method</span>

<span class="co"># 24 significantly differential genera</span>
<span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">multiple_group_anova</span><span class="op">)</span><span class="op">$</span><span class="va">feature</span>
<span class="va">markers</span>
<span class="co">#&gt;                        p__Bacteroidetes                         p__Unclassified </span>
<span class="co">#&gt;                      "p__Bacteroidetes"                       "p__Unclassified" </span>
<span class="co">#&gt;          p__Actinobacteria|g__Scardovia           p__Bacteroidetes|g__Alistipes </span>
<span class="co">#&gt;        "p__Actinobacteria|g__Scardovia"         "p__Bacteroidetes|g__Alistipes" </span>
<span class="co">#&gt;         p__Bacteroidetes|g__Bacteroides     p__Bacteroidetes|g__Parabacteroides </span>
<span class="co">#&gt;       "p__Bacteroidetes|g__Bacteroides"   "p__Bacteroidetes|g__Parabacteroides" </span>
<span class="co">#&gt;          p__Bacteroidetes|g__Prevotella              p__Firmicutes|g__Bulleidia </span>
<span class="co">#&gt;        "p__Bacteroidetes|g__Prevotella"            "p__Firmicutes|g__Bulleidia" </span>
<span class="co">#&gt;        p__Firmicutes|g__Catenibacterium              p__Firmicutes|g__Catonella </span>
<span class="co">#&gt;      "p__Firmicutes|g__Catenibacterium"            "p__Firmicutes|g__Catonella" </span>
<span class="co">#&gt;             p__Firmicutes|g__Holdemania          p__Firmicutes|g__Lactobacillus </span>
<span class="co">#&gt;           "p__Firmicutes|g__Holdemania"        "p__Firmicutes|g__Lactobacillus" </span>
<span class="co">#&gt;            p__Firmicutes|g__Macrococcus     p__Firmicutes|g__Peptostreptococcus </span>
<span class="co">#&gt;          "p__Firmicutes|g__Macrococcus"   "p__Firmicutes|g__Peptostreptococcus" </span>
<span class="co">#&gt;           p__Firmicutes|g__Ruminococcus            p__Firmicutes|g__Selenomonas </span>
<span class="co">#&gt;         "p__Firmicutes|g__Ruminococcus"          "p__Firmicutes|g__Selenomonas" </span>
<span class="co">#&gt;          p__Firmicutes|g__Streptococcus        p__Firmicutes|g__Subdoligranulum </span>
<span class="co">#&gt;        "p__Firmicutes|g__Streptococcus"      "p__Firmicutes|g__Subdoligranulum" </span>
<span class="co">#&gt;         p__Proteobacteria|g__Bartonella           p__Proteobacteria|g__Brucella </span>
<span class="co">#&gt;       "p__Proteobacteria|g__Bartonella"         "p__Proteobacteria|g__Brucella" </span>
<span class="co">#&gt;      p__Proteobacteria|g__Granulibacter     p__Proteobacteria|g__Rhodospirillum </span>
<span class="co">#&gt;    "p__Proteobacteria|g__Granulibacter"   "p__Proteobacteria|g__Rhodospirillum" </span>
<span class="co">#&gt;   p__Proteobacteria|g__Stenotrophomonas         p__Unclassified|g__Unclassified </span>
<span class="co">#&gt; "p__Proteobacteria|g__Stenotrophomonas"       "p__Unclassified|g__Unclassified"</span>
<span class="co"># take a marker "p__Bacteroidetes|g__Bacteroides"  </span>
<span class="co"># for example, we will show "p__Bacteroidetes|g__Bacteroides"  differ from </span>
<span class="co"># between Enterotype 2-Enterotype 1 and Enterotype 3-Enterotype 2.</span>
<span class="va">pht</span><span class="op">@</span><span class="va">result</span><span class="op">$</span><span class="st">"p__Bacteroidetes|g__Bacteroides"</span>
<span class="co">#&gt; DataFrame with 3 rows and 5 columns</span>
<span class="co">#&gt;                                       comparions  diff_mean      pvalue</span>
<span class="co">#&gt;                                      &lt;character&gt;  &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">#&gt; Enterotype 2-Enterotype 1 Enterotype 2-Enterot.. -0.2813948 4.77015e-08</span>
<span class="co">#&gt; Enterotype 3-Enterotype 1 Enterotype 3-Enterot.. -0.2604547 1.63635e-09</span>
<span class="co">#&gt; Enterotype 3-Enterotype 2 Enterotype 3-Enterot..  0.0209401 7.88993e-01</span>
<span class="co">#&gt;                             ci_lower   ci_upper</span>
<span class="co">#&gt;                            &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">#&gt; Enterotype 2-Enterotype 1 -0.3713469 -0.1914428</span>
<span class="co">#&gt; Enterotype 3-Enterotype 1 -0.3312286 -0.1896808</span>
<span class="co">#&gt; Enterotype 3-Enterotype 2 -0.0575765  0.0994567</span></code></pre></div>
<p>Visualization of post test result of a given feature.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># visualize the post hoc test result of Bacteroides</span>
<span class="fu"><a href="reference/plot_postHocTest.html">plot_postHocTest</a></span><span class="op">(</span><span class="va">pht</span>, feature <span class="op">=</span> <span class="st">"p__Bacteroidetes|g__Bacteroides"</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-plot-posthoctest-1.png"><!-- --></p>
</div>
</div>
<div id="metagenomeseq" class="section level2">
<h2 class="hasAnchor">
<a href="#metagenomeseq" class="anchor"></a>metagenomeSeq</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_mgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_metagenomeseq.html">run_metagenomeseq</a></span><span class="op">(</span>
  <span class="va">pediatric_ibd</span>,
  <span class="st">"Class"</span>,
  contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD"</span>,<span class="st">"Control"</span><span class="op">)</span>,
  pvalue_cutoff <span class="op">=</span> <span class="fl">0.1</span>,
  p_adjust <span class="op">=</span> <span class="st">"fdr"</span>
<span class="op">)</span>
<span class="va">mm_mgs</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ CSS ]</span>
<span class="co">#&gt; microbiome marker identity method: [ metagenomeSeq: ZILN ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 11 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 786 taxa and  43 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 43 samples by  2 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 786 taxa by 1 taxonomic ranks ]</span>

<span class="co"># multiple groups comparison</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html">subset_samples</a></span><span class="op">(</span>
  <span class="va">cid_ying</span>,
  <span class="va">Consistency</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"formed stool"</span>, <span class="st">"liquid"</span>, <span class="st">"semi-formed"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">mm_mgs_multiple</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_metagenomeseq.html">run_metagenomeseq</a></span><span class="op">(</span><span class="va">ps</span>, <span class="st">"Consistency"</span>, method <span class="op">=</span> <span class="st">"ZIG"</span><span class="op">)</span>
<span class="co">#&gt; it= 0, nll=608.38, log10(eps+1)=Inf, stillActive=669</span>
<span class="co">#&gt; it= 1, nll=621.65, log10(eps+1)=0.02, stillActive=137</span>
<span class="co">#&gt; it= 2, nll=617.53, log10(eps+1)=0.04, stillActive=132</span>
<span class="co">#&gt; it= 3, nll=613.29, log10(eps+1)=0.04, stillActive=124</span>
<span class="co">#&gt; it= 4, nll=608.76, log10(eps+1)=0.07, stillActive=95</span>
<span class="co">#&gt; it= 5, nll=605.16, log10(eps+1)=0.07, stillActive=60</span>
<span class="co">#&gt; it= 6, nll=603.49, log10(eps+1)=0.07, stillActive=35</span>
<span class="co">#&gt; it= 7, nll=604.06, log10(eps+1)=0.05, stillActive=23</span>
<span class="co">#&gt; it= 8, nll=607.70, log10(eps+1)=0.00, stillActive=0</span>
<span class="va">mm_mgs_multiple</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ CSS ]</span>
<span class="co">#&gt; microbiome marker identity method: [ metagenomeSeq: ZIG ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 486 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 669 taxa and  413 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 413 samples by  6 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 669 taxa by 1 taxonomic ranks ]</span></code></pre></div>
</div>
<div id="deseq2" class="section level2">
<h2 class="hasAnchor">
<a href="#deseq2" class="anchor"></a>DESeq2</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># two groups comparison</span>
<span class="va">mm_des</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_deseq2.html">run_deseq2</a></span><span class="op">(</span>
  <span class="va">pediatric_ibd</span>, 
  <span class="st">"Class"</span>, 
  contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"CD"</span><span class="op">)</span>, 
  pvalue_cutoff <span class="op">=</span> <span class="fl">0.05</span>, 
  p_adjust <span class="op">=</span> <span class="st">"fdr"</span>
<span class="op">)</span>
<span class="va">mm_des</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ RLE ]</span>
<span class="co">#&gt; microbiome marker identity method: [ DESeq2: Wald ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 47 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 786 taxa and  43 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 43 samples by  2 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 786 taxa by 1 taxonomic ranks ]</span>

<span class="co"># multiple groups</span>
<span class="va">mm_des_multiple</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_deseq2.html">run_deseq2</a></span><span class="op">(</span>
  <span class="va">ps</span>,
  <span class="st">"Consistency"</span>,
  pvalue_cutoff <span class="op">=</span> <span class="fl">0.05</span>,
  p_adjust <span class="op">=</span> <span class="st">"fdr"</span>
<span class="op">)</span>
<span class="va">mm_des_multiple</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ RLE ]</span>
<span class="co">#&gt; microbiome marker identity method: [ DESeq2: LRT ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 90 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 669 taxa and  413 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 413 samples by  6 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 669 taxa by 1 taxonomic ranks ]</span></code></pre></div>
</div>
<div id="edger" class="section level2">
<h2 class="hasAnchor">
<a href="#edger" class="anchor"></a>edgeR</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_edger</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_edger.html">run_edger</a></span><span class="op">(</span>
  <span class="va">pediatric_ibd</span>,
  <span class="st">"Class"</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD"</span>, <span class="st">"Control"</span><span class="op">)</span>,
  pvalue_cutoff <span class="op">=</span> <span class="fl">0.1</span>,
  p_adjust <span class="op">=</span> <span class="st">"fdr"</span>
<span class="op">)</span>
<span class="va">mm_edger</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ TMM ]</span>
<span class="co">#&gt; microbiome marker identity method: [ edgeR: LRT ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 34 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 786 taxa and  43 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 43 samples by  2 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 786 taxa by 1 taxonomic ranks ]</span>

<span class="co"># multiple groups</span>
<span class="va">mm_edger_multiple</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_edger.html">run_edger</a></span><span class="op">(</span>
  <span class="va">ps</span>,
  <span class="st">"Consistency"</span>,
  method  <span class="op">=</span> <span class="st">"QLFT"</span>,
  pvalue_cutoff <span class="op">=</span> <span class="fl">0.05</span>,
  p_adjust <span class="op">=</span> <span class="st">"fdr"</span>
<span class="op">)</span>
<span class="va">mm_edger_multiple</span>
<span class="co">#&gt; microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">#&gt; normalization method:              [ TMM ]</span>
<span class="co">#&gt; microbiome marker identity method: [ edgeR: QLFT ]</span>
<span class="co">#&gt; marker_table() Marker Table:       [ 325 microbiome markers with 5 variables ]</span>
<span class="co">#&gt; otu_table()    OTU Table:          [ 669 taxa and  413 samples ]</span>
<span class="co">#&gt; sample_data()  Sample Data:        [ 413 samples by  6 sample variables ]</span>
<span class="co">#&gt; tax_table()    Taxonomy Table:     [ 669 taxa by 1 taxonomic ranks ]</span></code></pre></div>
</div>
<div id="ancom" class="section level2">
<h2 class="hasAnchor">
<a href="#ancom" class="anchor"></a>ANCOM</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_ancom</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_ancom.html">run_ancom</a></span><span class="op">(</span><span class="va">ecam</span>, <span class="st">"delivery"</span>, p_adjust <span class="op">=</span> <span class="st">"none"</span>, theta <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span>
<span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">mm_ancom</span><span class="op">)</span>
<span class="co">#&gt;                                                                                                               feature</span>
<span class="co">#&gt; marker1                                                                                 k__Bacteria|p__Actinobacteria</span>
<span class="co">#&gt; marker2                                                               k__Bacteria|p__Actinobacteria|c__Actinobacteria</span>
<span class="co">#&gt; marker3                                          k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales</span>
<span class="co">#&gt; marker4                    k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae</span>
<span class="co">#&gt; marker5 k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium</span>
<span class="co">#&gt;         enrich_group ef_CLR_diff_mean   W</span>
<span class="co">#&gt; marker1     Cesarean        0.2903173 105</span>
<span class="co">#&gt; marker2     Cesarean        0.3047846 105</span>
<span class="co">#&gt; marker3     Cesarean        0.3759503 106</span>
<span class="co">#&gt; marker4     Cesarean        0.3759503 106</span>
<span class="co">#&gt; marker5     Cesarean        0.3759503 106</span></code></pre></div>
</div>
<div id="ancombc" class="section level2">
<h2 class="hasAnchor">
<a href="#ancombc" class="anchor"></a>ANCOMBC</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_ancombc</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_ancombc.html">run_ancombc</a></span><span class="op">(</span><span class="va">ecam</span>, <span class="st">"delivery"</span>, group <span class="op">=</span> <span class="st">"delivery"</span><span class="op">)</span>
<span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">mm_ancombc</span><span class="op">)</span>
<span class="co">#&gt;                                                                                                                 feature</span>
<span class="co">#&gt; marker1                                                                                   k__Bacteria|p__Actinobacteria</span>
<span class="co">#&gt; marker2                                                                                  k__Bacteria|p__Verrucomicrobia</span>
<span class="co">#&gt; marker3                                                                 k__Bacteria|p__Actinobacteria|c__Actinobacteria</span>
<span class="co">#&gt; marker4                                                             k__Bacteria|p__Proteobacteria|c__Betaproteobacteria</span>
<span class="co">#&gt; marker5                                                            k__Bacteria|p__Proteobacteria|c__Deltaproteobacteria</span>
<span class="co">#&gt; marker6                                                              k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae</span>
<span class="co">#&gt; marker7                                              k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Actinomycetales</span>
<span class="co">#&gt; marker8                                            k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales</span>
<span class="co">#&gt; marker9                                                         k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales</span>
<span class="co">#&gt; marker10                                         k__Bacteria|p__Proteobacteria|c__Betaproteobacteria|o__Burkholderiales</span>
<span class="co">#&gt; marker11                                     k__Bacteria|p__Proteobacteria|c__Deltaproteobacteria|o__Desulfovibrionales</span>
<span class="co">#&gt; marker12                                        k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pseudomonadales</span>
<span class="co">#&gt; marker13                                       k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales</span>
<span class="co">#&gt; marker14                         k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Actinomycetales|f__Actinomycetaceae</span>
<span class="co">#&gt; marker15                     k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae</span>
<span class="co">#&gt; marker16                                 k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae</span>
<span class="co">#&gt; marker17                                    k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Streptococcaceae</span>
<span class="co">#&gt; marker18                       k__Bacteria|p__Proteobacteria|c__Betaproteobacteria|o__Burkholderiales|f__Alcaligenaceae</span>
<span class="co">#&gt; marker19              k__Bacteria|p__Proteobacteria|c__Deltaproteobacteria|o__Desulfovibrionales|f__Desulfovibrionaceae</span>
<span class="co">#&gt; marker20                    k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pseudomonadales|f__Pseudomonadaceae</span>
<span class="co">#&gt; marker21                k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae</span>
<span class="co">#&gt; marker22          k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Actinomycetales|f__Actinomycetaceae|g__Actinomyces</span>
<span class="co">#&gt; marker23  k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium</span>
<span class="co">#&gt; marker24                   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae|g__Prevotella</span>
<span class="co">#&gt; marker25                   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Streptococcaceae|g__Streptococcus</span>
<span class="co">#&gt; marker26                  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Lachnospiraceae|g__[Ruminococcus]</span>
<span class="co">#&gt; marker27                     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Lachnospiraceae|g__Coprococcus</span>
<span class="co">#&gt; marker28                    k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Oscillospira</span>
<span class="co">#&gt; marker29                       k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Veillonellaceae|g__Dialister</span>
<span class="co">#&gt; marker30         k__Bacteria|p__Proteobacteria|c__Betaproteobacteria|o__Burkholderiales|f__Alcaligenaceae|g__Sutterella</span>
<span class="co">#&gt; marker31 k__Bacteria|p__Proteobacteria|c__Deltaproteobacteria|o__Desulfovibrionales|f__Desulfovibrionaceae|g__Bilophila</span>
<span class="co">#&gt; marker32     k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pseudomonadales|f__Pseudomonadaceae|g__Pseudomonas</span>
<span class="co">#&gt; marker33 k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae|g__Akkermansia</span>
<span class="co">#&gt;          enrich_group      ef_W       pvalue         padj</span>
<span class="co">#&gt; marker1      Cesarean -3.344472 0.0008243933 0.0008243933</span>
<span class="co">#&gt; marker2       Vaginal  2.039939 0.0413564022 0.0413564022</span>
<span class="co">#&gt; marker3      Cesarean -3.330633 0.0008664864 0.0008664864</span>
<span class="co">#&gt; marker4      Cesarean -2.294782 0.0217456056 0.0217456056</span>
<span class="co">#&gt; marker5      Cesarean -3.488293 0.0004861155 0.0004861155</span>
<span class="co">#&gt; marker6       Vaginal  2.039939 0.0413564022 0.0413564022</span>
<span class="co">#&gt; marker7       Vaginal  1.967299 0.0491488066 0.0491488066</span>
<span class="co">#&gt; marker8      Cesarean -3.518274 0.0004343629 0.0004343629</span>
<span class="co">#&gt; marker9      Cesarean -2.116808 0.0342761767 0.0342761767</span>
<span class="co">#&gt; marker10     Cesarean -2.294782 0.0217456056 0.0217456056</span>
<span class="co">#&gt; marker11     Cesarean -3.488293 0.0004861155 0.0004861155</span>
<span class="co">#&gt; marker12     Cesarean -2.259407 0.0238580789 0.0238580789</span>
<span class="co">#&gt; marker13      Vaginal  2.039939 0.0413564022 0.0413564022</span>
<span class="co">#&gt; marker14      Vaginal  2.391804 0.0167658080 0.0167658080</span>
<span class="co">#&gt; marker15     Cesarean -3.518274 0.0004343629 0.0004343629</span>
<span class="co">#&gt; marker16      Vaginal  3.674617 0.0002382063 0.0002382063</span>
<span class="co">#&gt; marker17     Cesarean -2.022194 0.0431563503 0.0431563503</span>
<span class="co">#&gt; marker18     Cesarean -2.294782 0.0217456056 0.0217456056</span>
<span class="co">#&gt; marker19     Cesarean -3.488293 0.0004861155 0.0004861155</span>
<span class="co">#&gt; marker20     Cesarean -2.231962 0.0256174686 0.0256174686</span>
<span class="co">#&gt; marker21      Vaginal  2.039939 0.0413564022 0.0413564022</span>
<span class="co">#&gt; marker22      Vaginal  2.391804 0.0167658080 0.0167658080</span>
<span class="co">#&gt; marker23     Cesarean -3.518274 0.0004343629 0.0004343629</span>
<span class="co">#&gt; marker24      Vaginal  3.674617 0.0002382063 0.0002382063</span>
<span class="co">#&gt; marker25     Cesarean -2.022194 0.0431563503 0.0431563503</span>
<span class="co">#&gt; marker26     Cesarean -1.988102 0.0468003786 0.0468003786</span>
<span class="co">#&gt; marker27      Vaginal  2.482640 0.0130412723 0.0130412723</span>
<span class="co">#&gt; marker28     Cesarean -2.141495 0.0322341803 0.0322341803</span>
<span class="co">#&gt; marker29      Vaginal  3.074502 0.0021085463 0.0021085463</span>
<span class="co">#&gt; marker30     Cesarean -2.294782 0.0217456056 0.0217456056</span>
<span class="co">#&gt; marker31     Cesarean -3.488293 0.0004861155 0.0004861155</span>
<span class="co">#&gt; marker32     Cesarean -2.231962 0.0256174686 0.0256174686</span>
<span class="co">#&gt; marker33      Vaginal  2.039939 0.0413564022 0.0413564022</span></code></pre></div>
</div>
<div id="superivised-learning" class="section level2">
<h2 class="hasAnchor">
<a href="#superivised-learning" class="anchor"></a>Superivised learning</h2>
<div id="logistic-regression" class="section level3">
<h3 class="hasAnchor">
<a href="#logistic-regression" class="anchor"></a>Logistic regression</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_lr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_sl.html">run_sl</a></span><span class="op">(</span><span class="va">enterotypes_arumugam</span>, <span class="st">"Gender"</span>, method <span class="op">=</span> <span class="st">"LR"</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: lattice</span>
<span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">mm_lr</span><span class="op">)</span>
<span class="co">#&gt;                                        feature enrich_group    ef_imp</span>
<span class="co">#&gt; marker1        p__Proteobacteria|g__Variovorax            F 100.00000</span>
<span class="co">#&gt; marker2         p__Proteobacteria|g__Aeromonas            F  86.21530</span>
<span class="co">#&gt; marker3        p__Proteobacteria|g__Aliivibrio            M  82.51265</span>
<span class="co">#&gt; marker4       p__Bacteroidetes|g__Zunongwangia            F  68.17224</span>
<span class="co">#&gt; marker5       p__Proteobacteria|g__Azotobacter            F  59.64652</span>
<span class="co">#&gt; marker6     p__Bacteroidetes|g__Capnocytophaga            M  52.47557</span>
<span class="co">#&gt; marker7  p__Proteobacteria|g__Methylobacterium            M  42.59024</span>
<span class="co">#&gt; marker8    p__Proteobacteria|g__Bradyrhizobium            M  37.50843</span>
<span class="co">#&gt; marker9    p__Proteobacteria|g__Photobacterium            M  33.71803</span>
<span class="co">#&gt; marker10      p__Proteobacteria|g__Simonsiella            F  33.41961</span></code></pre></div>
</div>
<div id="random-forest" class="section level3">
<h3 class="hasAnchor">
<a href="#random-forest" class="anchor"></a>Random forest</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># must specify the importance</span>
<span class="va">mm_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_sl.html">run_sl</a></span><span class="op">(</span>
  <span class="va">enterotypes_arumugam</span>, 
  <span class="st">"Gender"</span>, 
  method <span class="op">=</span> <span class="st">"RF"</span>, 
  importance <span class="op">=</span> <span class="st">"impurity"</span>
<span class="op">)</span>
<span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">mm_rf</span><span class="op">)</span>
<span class="co">#&gt;                                            feature enrich_group    ef_imp</span>
<span class="co">#&gt; marker1               p__Firmicutes|g__Megasphaera            F 100.00000</span>
<span class="co">#&gt; marker2           p__Firmicutes|g__Ruminococcaceae            F  70.45596</span>
<span class="co">#&gt; marker3  p__Proteobacteria|g__Escherichia/Shigella            F  69.31432</span>
<span class="co">#&gt; marker4        p__Firmicutes|g__Desulfitobacterium            F  69.01567</span>
<span class="co">#&gt; marker5        p__Firmicutes|g__Peptostreptococcus            M  66.95028</span>
<span class="co">#&gt; marker6           p__Firmicutes|g__Subdoligranulum            F  65.87990</span>
<span class="co">#&gt; marker7          p__Bacteroidetes|g__Porphyromonas            F  65.41748</span>
<span class="co">#&gt; marker8           p__Firmicutes|g__Lachnospiraceae            F  65.10861</span>
<span class="co">#&gt; marker9             p__Firmicutes|g__Anaerotruncus            F  64.84360</span>
<span class="co">#&gt; marker10   p__Proteobacteria|g__Enterobacteriaceae            F  64.63217</span></code></pre></div>
</div>
<div id="support-vector-machine" class="section level3">
<h3 class="hasAnchor">
<a href="#support-vector-machine" class="anchor"></a>Support vector machine</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_svm</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_sl.html">run_sl</a></span><span class="op">(</span><span class="va">enterotypes_arumugam</span>, <span class="st">"Gender"</span>, method <span class="op">=</span> <span class="st">"SVM"</span><span class="op">)</span>
<span class="fu"><a href="reference/marker_table.html">marker_table</a></span><span class="op">(</span><span class="va">mm_svm</span><span class="op">)</span>
<span class="co">#&gt;                                            feature enrich_group    ef_imp</span>
<span class="co">#&gt; marker1        p__Firmicutes|g__Peptostreptococcus            M 100.00000</span>
<span class="co">#&gt; marker2  p__Proteobacteria|g__Escherichia/Shigella            F  97.81022</span>
<span class="co">#&gt; marker3          p__Firmicutes|g__Faecalibacterium            F  96.35036</span>
<span class="co">#&gt; marker4             p__Firmicutes|g__Clostridiales            F  93.43066</span>
<span class="co">#&gt; marker5             p__Firmicutes|g__Anaerotruncus            F  91.24088</span>
<span class="co">#&gt; marker6               p__Firmicutes|g__Mitsuokella            M  86.86131</span>
<span class="co">#&gt; marker7    p__Proteobacteria|g__Enterobacteriaceae            F  82.48175</span>
<span class="co">#&gt; marker8           p__Firmicutes|g__Acidaminococcus            F  78.83212</span>
<span class="co">#&gt; marker9           p__Proteobacteria|g__Citrobacter            F  78.10219</span>
<span class="co">#&gt; marker10          p__Proteobacteria|g__Haemophilus            M  74.45255</span></code></pre></div>
</div>
</div>
<div id="visualzatiton" class="section level2">
<h2 class="hasAnchor">
<a href="#visualzatiton" class="anchor"></a>Visualzatiton</h2>
<div id="cladogram-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#cladogram-plot" class="anchor"></a>Cladogram plot</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plot_cladogram.html">plot_cladogram</a></span><span class="op">(</span><span class="va">mm_lefse</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-cladogram-1.png"><!-- --></p>
<p>It’s recommended to use a named vector to set the colors of enriched group:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plot_cladogram.html">plot_cladogram</a></span><span class="op">(</span><span class="va">mm_lefse</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Healthy <span class="op">=</span> <span class="st">"blue"</span>, Tumor <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-cladogram-color-1.png"><!-- --></p>
</div>
</div>
<div id="welcome" class="section level2">
<h2 class="hasAnchor">
<a href="#welcome" class="anchor"></a>Welcome</h2>
<p><strong>microbiomeMarker is still a newborn, and only contains lefse methods right now. Your suggestion and contribution will be highly appreciated.</strong></p>
</div>
<div id="acknowledgement" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgement" class="anchor"></a>Acknowledgement</h2>
<ul>
<li>
<a href="https://bitbucket.org/biobakery/biobakery/wiki/lefse">lefse python script</a>, The main lefse code are translated from <strong>lefse python script</strong>,</li>
<li>
<a href="https://github.com/lch14forever/microbiomeViz">microbiomeViz</a>, cladogram visualization of lefse is modified from <strong>microbiomeViz</strong>.</li>
<li>
<a href="https://github.com/joey711/phyloseq">phyloseq</a>, the main data structures used in <strong>microbiomeMarker</strong> are from or inherit from <code>phyloseq-class</code> in package <strong>phyloseq</strong>.</li>
<li>
<a href="https://github.com/YuLab-SMU/MicrobiotaProcess">MicrobiotaProcess</a>, function <code><a href="reference/import_dada2.html">import_dada2()</a></code> and <code><a href="reference/import_qiime2.html">import_qiime2()</a></code> are modified from the <code>MicrobiotaProcess::import_dada2()</code>.</li>
<li>
<a href="https://github.com/jbisanz/qiime2R">qiime2R</a>, <code><a href="reference/import_qiime2.html">import_qiime2()</a></code> are refer to the functions in qiime2R.</li>
</ul>
</div>
<div id="question" class="section level2">
<h2 class="hasAnchor">
<a href="#question" class="anchor"></a>Question</h2>
<p>If you have any question, please file an issue on the issue tracker following the instructions in the issue template:</p>
<p>Please briefly describe your problem, what output actually happened, and what output you expect.</p>
<p>Please provide a minimal reproducible example. For more details on how to make a great minimal reproducible example, see <a href="https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example" class="uri">https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example</a> and <a href="https://www.tidyverse.org/help/#reprex" class="uri">https://www.tidyverse.org/help/#reprex</a>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>Brief description of the problem</span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="co"># insert minimal reprducible example here</span></span></code></pre></div>
</div>
<div id="reference" class="section level2">
<h2 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Love_2014" class="csl-entry">
<p>Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.” <em>Genome Biology</em> 15 (12). <a href="https://doi.org/10.1186/s13059-014-0550-8" class="uri">https://doi.org/10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-Parks_2014" class="csl-entry">
<p>Parks, Donovan H., Gene W. Tyson, Philip Hugenholtz, and Robert G. Beiko. 2014. “STAMP: Statistical Analysis of Taxonomic and Functional Profiles.” <em>Bioinformatics</em> 30 (21): 3123–24. <a href="https://doi.org/10.1093/bioinformatics/btu494" class="uri">https://doi.org/10.1093/bioinformatics/btu494</a>.</p>
</div>
<div id="ref-Paulson_2013" class="csl-entry">
<p>Paulson, Joseph N, O Colin Stine, H’ector Corrada Bravo, and Mihai Pop.</p>
<ol>
<li>“Differential Abundance Analysis for Microbial Marker-Gene Surveys.” <em>Nature Methods</em> 10 (12): 1200–1202. <a href="https://doi.org/10.1038/nmeth.2658" class="uri">https://doi.org/10.1038/nmeth.2658</a>.</li>
</ol>
</div>
<div id="ref-Robinson_2009" class="csl-entry">
<p>Robinson, M. D., D. J. McCarthy, and G. K. Smyth. 2009. “<span class="nocase">edgeR</span>: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616" class="uri">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
<div id="ref-Segata_2011" class="csl-entry">
<p>Segata, Nicola, Jacques Izard, Levi Waldron, Dirk Gevers, Larisa Miropolsky, Wendy S Garrett, and Curtis Huttenhower. 2011. “Metagenomic Biomarker Discovery and Explanation.” <em>Genome Biology</em> 12 (6): R60. <a href="https://doi.org/10.1186/gb-2011-12-6-r60" class="uri">https://doi.org/10.1186/gb-2011-12-6-r60</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/yiluheihei/microbiomeMarker/">https://​github.com/​yiluheihei/​microbiomeMarker/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/yiluheihei/microbiomeMarker/issues">https://​github.com/​yiluheihei/​microbiomeMarker/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Yang Cao <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Yang Cao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
