---
title:: R package for microbiome biomarker discovery
bibliography: references.bib
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r set, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)

# knitciations
library(knitcitations)
cleanbib()
options("citation_format" = "pandoc")
```

<!-- badges: start -->
[![R build status](https://github.com/yiluheihei/microbiomeMarker/workflows/R-CMD-check/badge.svg)](https://github.com/yiluheihei/microbiomeMarker/actions)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://github.com/yiluheihei/microbiomeMarker/blob/master/LICENSE.md)
[![Codecov test coverage](https://codecov.io/gh/yiluheihei/microbiomeMarker/branch/master/graph/badge.svg)](https://codecov.io/gh/yiluheihei/microbiomeMarker?branch=master)
[![DOI](https://zenodo.org/badge/215731961.svg)](https://zenodo.org/badge/latestdoi/215731961)
<!-- badges: end -->

## Motivation

**The aim of this package is to build a unified toolbox in R for mcirobiome 
biomarker discovery by integrating various existing methods.**

Many statistical methods have been proposed to discovery the microbiome biomaker
by compare the taxon abundance between different classes. Some methods developed 
specifically for microbial community, such as  linear discriminant analysis 
(LDA) effect size (LEfSe) `r citep("10.1186/gb-2011-12-6-r60")`, metagenomeSeq 
`r citep("10.1038/nmeth.2658")`; and some methods developed specifically for 
RNA-Seq data, such as DESeq2 `r citep("10.1186/s13059-014-0550-8")` and 
edgeR `r citep("10.1093/bioinformatics/btp616")`, have been proposed for 
microbiome biomarker discovery. We usually use several methods for microbiome 
biomarker discovery and compare the results, which requires multiple tools 
developed in different programming, even in different OS.

**microbiomeMarker** take the `phyloseq-class` object in package 
[phyloseq](https://github.com/joey711/phyloseq) as input, since **phyloseq** is 
the most popular R package in microbiome analysis and with phyloseq 
you can easily import taxon abundance and phylogenetic tree of taxon output from 
common microbiome bioinformatics platforms, such as 
[DADA2](https://benjjneb.github.io/dada2/) and [qiime2](https://qiime2.org/).

## Installation

You can install the package directly from github

```{r install, eval=FALSE}
if (!require(remotes)) install.packages("remotes")
remotes::install_github("yiluheihei/microbiomeMarker")
```

## LEfSe

Curently, LEfSe is the most used tool for microbiome biomarker discovery, and 
the first method to integrate to **microbiomeMarker** is LEfSe.

### lefse analysis

```{r lefse,warning=FALSE}
library(microbiomeMarker)
library(ggplot2)

# sample data from lefse python script. The dataset contains 30 abundance 
# profiles (obtained processing the 16S reads with RDP) belonging to 10 rag2 
# (control) and 20 truc (case) mice
data("spontaneous_colitis")
lefse_out <- lefse(
  spontaneous_colitis, 
  normalization = 1e6, 
  class = "class", 
  multicls_strat = TRUE
)
# lefse return a microbioMarker class inherits from phyloseq
lefse_out
```

The microbiome biomarker information was stored in a new data structure 
`marker_table-class` inherited from `data.frame`, and you can access it by 
using `marker_table()`.

```{r}
head(marker_table(lefse_out))
```

### Visualization of the result of lefse analysis

Bar plot for output of lefse:

```{r lefse-barplot,dpi=300}
lefse_barplot(lefse_out, label_level = 1) +
  scale_fill_manual(values = c("rag2" = "blue", "truc" = "red"))
```

Cladogram plot for output of lefse

```{r lefse-cladogram,dpi=300}
lefse_cladogram(lefse_out, color = c("blue", "red"))
```

## statistical analysis (stamp)

STAMP `r citep("10.1093/bioinformatics/btu494")` is a widely-used graphical 
software package that provides "best pratices" in choose appropriate statistical
methods for microbial taxonomic and functional analysis. Users can tests for 
both two groups or multiple groups, and effect sizes and confidence intervals are
supported that allows critical assessment of the biological relevancy of test 
results. Here, **microbiomeMarker** also integrates the statistical methods used
in STAMP for microbial comparison analysis between two-groups and multiple-groups.

### Statitical analysis between two groups

Function `test_two_groups()` is developed for statistical test between two 
groups, and three test methods are provided: welch test, t test and white test.

```{r test-two-groups}
data("enterotypes_arumugam")
# take welch test for example
two_group_welch <- test_two_groups(
  enterotypes_arumugam, 
  group = "Gender", 
  rank_name = "Genus", 
  method = "welch.test"
)

# three significantly differential genera (marker)
two_group_welch
# details of result of the three markers
head(marker_table(two_group_welch))
```

### Statistical analysis multiple groups

Function `test_multiple_groups()` is constructed for statistical test for multiple
groups, two test method are provided: anova and kruskal test.

```{r test-multiple-groups}
# three groups
ps <- phyloseq::subset_samples(
  enterotypes_arumugam,
  Enterotype %in% c("Enterotype 3", "Enterotype 2", "Enterotype 1")
)

multiple_group_anova <-  test_multiple_groups(
  ps, 
  group = "Enterotype", 
  rank_name = "Genus",
  method = "anova"
)

# 22 markers (significantly differential genera)
multiple_group_anova
head(marker_table(multiple_group_anova))
```

The result of multiple group statistic specified whether the means of all groups
is equal or not. To identify which pairs of groups may differ from each other, 
post-hoc test must be performed.

```{r post-hoc-test}
pht <- posthoc_test(ps, group = "Enterotype" , rank_name = "Genus")
pht

# 22 significantly differential genera
markers <- marker_table(multiple_group_anova)$feature
markers
# take a marker Bacteroides for example, we will show Bacteroides differ from 
# between Enterotype 2-Enterotype 1 and Enterotype 3-Enterotype 2.
pht@result$Bacteroides
```

Visualization of post test result of a given feature.

```{r plot-posthoctest}
# visualize the post hoc test result of Bacteroides
plot_postHocTest(pht, feature = "Bacteroides")
```




## Welcome

**microbiomeMarker is still a newborn, and only contains lefse methods right now. 
Your suggestion and contribution will be highly appreciated.**

## Citation

Kindly cite as follows:  Yang Cao (2020). microbiomeMarker: microbiome biomarker
analysis. R package version 0.0.1.9000. https://github.com/yiluheihei/microbiomeMarker. 
DOI: [10.5281/zenodo.3749415](https://doi.org/10.5281/zenodo.3749415).

## Acknowledgement

- [lefse python script](https://bitbucket.org/biobakery/biobakery/wiki/lefse), 
  The main lefse code are translated from **lefse python script**,
- [microbiomeViz](https://github.com/lch14forever/microbiomeViz), cladogram 
  visualization of lefse is modified from **microbiomeViz**.
- [phyloseq](https://github.com/joey711/phyloseq), the main data structures used 
  in **microbiomeMarker** are from or inherit from `phyloseq-class` in package 
  **phyloseq**.

## Question

If you have any question, please file an issue on the issue tracker following 
the instructions in the issue template:

Please briefly describe your problem, what output actually happend, and what 
output you expect.

Please provide a minimal reproducible example. For more deails on how to make 
a great minimal reproducible example, see https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example and
https://www.tidyverse.org/help/#reprex.

```
Brief description of the problem

# insert minimal reprducible example here
```

## Reference

```{r reference, echo=FALSE, message=FALSE}
knitcitations::write.bibtex(file = "references.bib")
```

