---
title:: R package for microbiome biomarker discovery
bibliography: references.bib
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r set, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)

# knitciations
# library(knitcitations)
# cleanbib()
# options("citation_format" = "pandoc")
```

<!-- badges: start -->
[![R build status](https://github.com/yiluheihei/microbiomeMarker/workflows/R-CMD-check/badge.svg)](https://github.com/yiluheihei/microbiomeMarker/actions)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://github.com/yiluheihei/microbiomeMarker/blob/master/LICENSE.md)
[![Codecov test coverage](https://codecov.io/gh/yiluheihei/microbiomeMarker/branch/master/graph/badge.svg)](https://codecov.io/gh/yiluheihei/microbiomeMarker?branch=master)
[![DOI](https://zenodo.org/badge/215731961.svg)](https://zenodo.org/badge/latestdoi/215731961)
<!-- badges: end -->

**If you think this project is helpful to you, you can give this project a star** :star:

## Motivation

**The aim of this package is to build a unified toolbox in R for mcirobiome 
biomarker discovery by integrating various existing methods.**

Many statistical methods have been proposed to discovery the microbiome biomaker
by compare the taxon abundance between different classes. Some methods developed 
specifically for microbial community, such as  linear discriminant analysis 
(LDA) effect size (LEfSe) [@Segata_2011], metagenomeSeq [@Paulson_2013]; and 
some methods developed specifically for RNA-Seq data, such as DESeq2 
[@Love_2014] and edgeR [@{Robinson_2009], have been proposed for microbiome 
biomarker discovery. We usually use several methods for microbiome biomarker 
discovery and compare the results, which requires multiple tools developed in 
different programming, even in different OS.

**microbiomeMarker** take the `phyloseq-class` object in package 
[phyloseq](https://github.com/joey711/phyloseq) as input, since **phyloseq** is 
the most popular R package in microbiome analysis and with phyloseq 
you can easily import taxon abundance and phylogenetic tree of taxon output from 
common microbiome bioinformatics platforms, such as 
[DADA2](https://benjjneb.github.io/dada2/) and [qiime2](https://qiime2.org/).

## Installation

You can install the package directly from github

```{r install, eval=FALSE}
if (!require(remotes)) install.packages("remotes")
remotes::install_github("yiluheihei/microbiomeMarker")
```

## Data import

Since  [phyloseq](https://github.com/joey711/phyloseq) objects are a great 
data-standard for microbiome data in R, the core functions in 
**microbiomeMarker** take `phylosq` object as input. Conveniently, 
**microbiomeMarker** provides features to import external data files form two 
common tools of microbiome analysis, [qiime2](http://qiime.org/) and
[dada2](https://benjjneb.github.io/dada2).

### Import from dada2

The output of the [dada2](https://benjjneb.github.io/dada2) pipeline is a 
feature table of amplicon sequence variants (an ASV table): A matrix with rows 
corresponding to samples and columns to ASVs, in which the value of each entry 
is the number of times that ASV was observed in that sample. This table is 
analogous to the traditional OTU table. Conveniently, taxa names are saved as 

```{r import-dada2}
library(microbiomeMarker)

seq_tab <- readRDS(system.file("extdata", "dada2_seqtab.rds",
  package= "microbiomeMarker"))
tax_tab <- readRDS(system.file("extdata", "dada2_taxtab.rds",
 package= "microbiomeMarker"))
sam_tab <- read.table(system.file("extdata", "dada2_samdata.txt",
 package= "microbiomeMarker"), sep = "\t", header = TRUE, row.names = 1)
ps <- import_dada2(seq_tab = seq_tab, tax_tab = tax_tab, sam_tab = sam_tab)
ps
```

### Import from qiime2

[qiime2](http://qiime.org/) is the most widely used software for metagenomic
analysis. User can import the feature table, taxonomic table, phylogenetic 
tree, representative sequence and sample metadata from qiime2 using
`import_qiime2()`.

```{r import-qiime2}
otuqza_file <- system.file("extdata", "table.qza",package = "microbiomeMarker")
taxaqza_file <- system.file("extdata", "taxonomy.qza",package = "microbiomeMarker")
sample_file <- system.file(
  "extdata", "sample-metadata.tsv",
  package = "microbiomeMarker"
)
treeqza_file <- system.file("extdata", "tree.qza",package = "microbiomeMarker")
ps <- import_qiime2(
  otu_qza = otuqza_file, taxa_qza = taxaqza_file,
  sam_tab = sample_file, tree_qza = treeqza_file
)
ps
```

### Other import functions reexport from phyloseq

**microbiomeMarker** reexports three import functions from **phyloseq**, 
including `import_biom()`, `import_qiime()` and `import_mothur()`, to help users
to import data from [biom file](http://biom-format.org/), and output from 
[qiime](http://www.qiime.org/) and [mothur](http://www.mothur.org/). More 
details on these three import functions can be see from
[here](https://joey711.github.io/phyloseq/import-data.html#the_import_family_of_functions).

Users can also import the external files into `phyloseq` object manually. For 
more details on how to create `phyloseq` object from manually imported data, 
please see 
[this tutorial](http://joey711.github.io/phyloseq/import-data.html#manual).

## LEfSe

Curently, LEfSe is the most used tool for microbiome biomarker discovery, and 
the first method to integrate to **microbiomeMarker** is LEfSe.

### lefse analysis

```{r lefse,warning=FALSE}
library(ggplot2)

# sample data from lefse python script. The dataset contains 30 abundance 
# profiles (obtained processing the 16S reads with RDP) belonging to 10 rag2 
# (control) and 20 truc (case) mice
data("spontaneous_colitis")
# add prefix of ranks
spontaneous_colitis <- microbiomeMarker:::add_prefix_summarized(
  spontaneous_colitis,
  c("k", "p", "c", "o", "f", "g")
)
mm <- lefse(
  spontaneous_colitis, 
  normalization = 1e6, 
  class = "class", 
  multicls_strat = TRUE
)
# lefse return a microbioMarker class inherits from phyloseq
mm
```

The microbiome biomarker information was stored in a new data structure 
`marker_table-class` inherited from `data.frame`, and you can access it by 
using `marker_table()`.

```{r}
head(marker_table(mm))
```

### Visualization of the result of lefse analysis

Bar plot for output of lefse:

```{r lefse-barplot,dpi=300}
lefse_barplot(mm, label_level = 1) +
  scale_fill_manual(values = c("rag2" = "blue", "truc" = "red"))
```

## statistical analysis (stamp)

STAMP [@Parks_2014] is a widely-used graphical software package that provides 
"best pratices" in choose appropriate statisticalmethods for microbial taxonomic 
and functional analysis. Users can tests for both two groups or multiple groups, 
and effect sizes and confidence intervals are supported that allows critical 
assessment of the biological relevancy of test results. Here, 
**microbiomeMarker** also integrates the statistical methods used in STAMP for 
microbial comparison analysis between two-groups and multiple-groups.

### Statitical analysis between two groups

Function `test_two_groups()` is developed for statistical test between two 
groups, and three test methods are provided: welch test, t test and white test.

```{r test-two-groups}
data("enterotypes_arumugam")
# take welch test for example
two_group_welch <- test_two_groups(
  enterotypes_arumugam, 
  group = "Gender", 
  method = "welch.test"
)

# three significantly differential genera (marker)
two_group_welch
# details of result of the three markers
head(marker_table(two_group_welch))
```

### Statistical analysis multiple groups

Function `test_multiple_groups()` is constructed for statistical test for multiple
groups, two test method are provided: anova and kruskal test.

```{r test-multiple-groups}
# three groups
ps <- phyloseq::subset_samples(
  enterotypes_arumugam,
  Enterotype %in% c("Enterotype 3", "Enterotype 2", "Enterotype 1")
)

multiple_group_anova <-  test_multiple_groups(
  ps,
  group = "Enterotype", 
  method = "anova"
)

# 22 markers (significantly differential genera)
multiple_group_anova
head(marker_table(multiple_group_anova))
```

The result of multiple group statistic specified whether the means of all groups
is equal or not. To identify which pairs of groups may differ from each other, 
post-hoc test must be performed.

```{r post-hoc-test}
pht <- posthoc_test(ps, group = "Enterotype" , rank_name = "Genus")
pht

# 22 significantly differential genera
markers <- marker_table(multiple_group_anova)$feature
markers
# take a marker Bacteroides for example, we will show Bacteroides differ from 
# between Enterotype 2-Enterotype 1 and Enterotype 3-Enterotype 2.
pht@result$Bacteroides
```

Visualization of post test result of a given feature.

```{r plot-posthoctest}
# visualize the post hoc test result of Bacteroides
plot_postHocTest(pht, feature = "Bacteroides")
```

## Visulatiton

### Cladogram plot

```{r cladogram,dpi=300,fig.width=14}
plot_cladogram(mm, color = c("blue", "red"))
```

It's recommended to use a named vector to set the colors of enriched group:

```{r cladogram-color,dpi=300,fig.width=14}
plot_cladogram(mm, color = c(truc = "blue", rag2 = "red"))
```

## Welcome

**microbiomeMarker is still a newborn, and only contains lefse methods right now. 
Your suggestion and contribution will be highly appreciated.**

## Citation

Kindly cite as follows:  Yang Cao (2020). microbiomeMarker: microbiome biomarker
analysis. R package version 0.0.1.9000. https://github.com/yiluheihei/microbiomeMarker. 
DOI: [10.5281/zenodo.3749415](https://doi.org/10.5281/zenodo.3749415).

## Acknowledgement

- [lefse python script](https://bitbucket.org/biobakery/biobakery/wiki/lefse), 
  The main lefse code are translated from **lefse python script**,
- [microbiomeViz](https://github.com/lch14forever/microbiomeViz), cladogram 
  visualization of lefse is modified from **microbiomeViz**.
- [phyloseq](https://github.com/joey711/phyloseq), the main data structures used 
  in **microbiomeMarker** are from or inherit from `phyloseq-class` in package 
  **phyloseq**.
- [MicrobiotaProcess](https://github.com/YuLab-SMU/MicrobiotaProcess), function 
  `import_dada2()` and `import_qiime2()` are modified from the 
  `MicrobiotaProcess::import_dada2()`.
- [qiime2R](https://github.com/jbisanz/qiime2R), `import_qiime2()` are refer to 
  the functions in qiime2R.

## Question

If you have any question, please file an issue on the issue tracker following 
the instructions in the issue template:

Please briefly describe your problem, what output actually happened, and what 
output you expect.

Please provide a minimal reproducible example. For more details on how to make 
a great minimal reproducible example, see https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example and
https://www.tidyverse.org/help/#reprex.

```
Brief description of the problem

# insert minimal reprducible example here
```

## Reference

<div id="refs"></div>

