<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Comparing the results of differential analysis methods by Empirical power
and False Discovery Rate — compare_DA • microbiomeMarker</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Comparing the results of differential analysis methods by Empirical power
and False Discovery Rate — compare_DA"><meta property="og:description" content="Calculating power, false discovery rates, false positive rates and auc (
area under the receiver operating characteristic (ROC) curve)
for various DA methods."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microbiomeMarker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/microbiomeMarker-vignette.html">Tools for microbiome marker identification</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/yiluheihei/microbiomeMarker/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Comparing the results of differential analysis methods by Empirical power
and False Discovery Rate</h1>
    <small class="dont-index">Source: <a href="https://github.com/yiluheihei/microbiomeMarker/blob/HEAD/R/DA-comparing.R" class="external-link"><code>R/DA-comparing.R</code></a></small>
    <div class="hidden name"><code>compare_DA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Calculating power, false discovery rates, false positive rates and auc (
area under the receiver operating characteristic (ROC) curve)
for various DA methods.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">compare_DA</span><span class="op">(</span>
  <span class="va">ps</span>,
  <span class="va">group</span>,
  taxa_rank <span class="op">=</span> <span class="st">"none"</span>,
  <span class="va">methods</span>,
  args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  n_rep <span class="op">=</span> <span class="fl">20</span>,
  effect_size <span class="op">=</span> <span class="fl">5</span>,
  k <span class="op">=</span> <span class="cn">NULL</span>,
  relative <span class="op">=</span> <span class="cn">TRUE</span>,
  BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span>progressbar <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>ps, group, taxa_rank</dt>
<dd><p>main arguments of all differential analysis
methods. <code>ps</code>: a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq::phyloseq</a></code> object; <code>group</code>, character,
the variable to set the group, must be one of the var of the sample
metadata; <code>taxa_rank</code>: character, taxonomic rank, please not that <strong>since
the abundance table is spiked in the lowest level, only
<code>taxa_rank = "none"</code> is allowed</strong>.</p></dd>
<dt>methods</dt>
<dd><p>character vector, differential analysis methods to be
compared, available methods are "aldex", "ancom", "ancombc", "deseq2",
"edger", "lefse", "limma_voom", "metagenomeseq", "simple_stat".</p></dd>
<dt>args</dt>
<dd><p>named list, which used to set the extra arguments of the
differential analysis methods, so the names must be contained in <code>methods</code>.
For more see details below.</p></dd>
<dt>n_rep</dt>
<dd><p>integer, number of times to run the differential analyses.</p></dd>
<dt>effect_size</dt>
<dd><p>numeric, the effect size for the spike-ins. Default 5.</p></dd>
<dt>k</dt>
<dd><p>numeric vector of length 3, number of features to spike in each
tertile (lower, mid, upper), e.g. <code>k=c(5,10,15)</code> means 5 features spiked
in low abundance tertile, 10 features spiked in mid abundance tertile and
15 features spiked in high abundance tertile. Default <code>NULL</code>, which will
spike 2 percent of the total amount of features in each tertile (a total
of 6 percent), but minimum c(5,5,5).</p></dd>
<dt>relative</dt>
<dd><p>logical, whether rescale the total number of individuals
observed for each sample to the original level after spike-in. Default
<code>TRUE</code>.</p></dd>
<dt>BPPARAM</dt>
<dd><p><code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">BiocParallel::BiocParallelParam</a></code> instance defining the
parallel back-end.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>an <code>compareDA</code> object, which contains a two-length list of:</p><ul><li><p><code>metrics</code>: <code>data.frame</code>, FPR, AUC and spike detection rate for each run.</p></li>
<li><p><code>mm</code>: differential analysis results.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>To make this function support for different arguments for a certain DA method
<code>args</code> allows list of list of list e.g. <code>args = list(lefse = list(list(norm = "CPM"), list(norm = "TSS")))</code>, which specify to compare the different norm
arguments for lefse analysis.</p>
<p>For <code>taxa_rank</code>, only <code>taxa_rank = "none"</code> is supported, if this argument is
not "none", it will be forced to "none" internally.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yang Cao.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

