<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Import function to read the output of picrust2 as phyloseq object — import_picrust2 • microbiomeMarker</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Import function to read the output of picrust2 as phyloseq object — import_picrust2"><meta property="og:description" content="Import the output of picrust2 into phyloseq object"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microbiomeMarker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/microbiomeMarker-vignette.html">Tools for microbiome marker identification</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/yiluheihei/microbiomeMarker/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import function to read the output of picrust2 as phyloseq object</h1>
    <small class="dont-index">Source: <a href="https://github.com/yiluheihei/microbiomeMarker/blob/HEAD/R/import-picrust2.R" class="external-link"><code>R/import-picrust2.R</code></a></small>
    <div class="hidden name"><code>import_picrust2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Import the output of picrust2 into phyloseq object</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">import_picrust2</span><span class="op">(</span>
  <span class="va">feature_tab</span>,
  sam_tab <span class="op">=</span> <span class="cn">NULL</span>,
  trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PATHWAY"</span>, <span class="st">"COG"</span>, <span class="st">"EC"</span>, <span class="st">"KO"</span>, <span class="st">"PFAM"</span>, <span class="st">"TIGRFAM"</span>, <span class="st">"PHENO"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>feature_tab</dt>
<dd><p>character, file path of the prediction abundance table of
functional feature.</p></dd>


<dt>sam_tab</dt>
<dd><p>character, file path of the sample meta data.</p></dd>


<dt>trait</dt>
<dd><p>character, options are picrust2 function traits (including
"COG", "EC", "KO", "PFAM", "TIGRFAM", and "PHENO") and "PATHWAY".</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq::phyloseq</a></code> object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><a href="https://huttenhower.sph.harvard.edu/picrust/" class="external-link">PICRUSt2</a> is a software for
predicting abundances of functional profiles based on marker gene sequencing
data. The functional profiles can be predicted from the taxonomic
profiles using PICRUSt2. "Function" usually refers to gene families such as
KEGG orthologs and Enzyme Classification numbers, but predictions can be
made for any arbitrary trait.</p>
<p>In the <code>phyloseq object</code>, the predicted function abundance profile is stored
in <code>otu_table</code> slot. And the functional trait is saved in <code>tax_table</code> slot,
if the descriptions of function features is not added to the predicted table,
<code>tax_table</code> will have only one rank <code>Picrust_trait</code> to represent the function
feature id, or if the desciptions are added one more rank
<code>Picrust_description</code> will be added to represent the description of
function feature.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">sam_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span class="r-in">    <span class="st">"extdata"</span>, <span class="st">"picrust2_metadata.tsv"</span>,</span>
<span class="r-in">    package <span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">feature_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span class="r-in">    <span class="st">"extdata"</span>, <span class="st">"path_abun_unstrat_descrip.tsv.gz"</span>,</span>
<span class="r-in">    package <span class="op">=</span> <span class="st">"microbiomeMarker"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu">import_picrust2</span><span class="op">(</span><span class="va">feature_tab</span>, <span class="va">sam_tab</span>, trait <span class="op">=</span> <span class="st">"PATHWAY"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">ps</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 195 taxa and 24 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 24 samples by 3 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 195 taxa by 2 taxonomic ranks ]</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yang Cao.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer></div>

  


  

  </body></html>

