<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import function to read the the output of dada2 as phyloseq object — import_dada2 • microbiomeMarker</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Import function to read the the output of dada2 as phyloseq object — import_dada2" />
<meta property="og:description" content="Import the output of dada2 into phyloseq object" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microbiomeMarker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/microbiomeMarker-vignette.html">Tools for microbiome marker identification</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/yiluheihei/microbiomeMarker/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import function to read the the output of dada2 as phyloseq object</h1>
    <small class="dont-index">Source: <a href='https://github.com/yiluheihei/microbiomeMarker/blob/master/R/import-dada2.R'><code>R/import-dada2.R</code></a></small>
    <div class="hidden name"><code>import_dada2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Import the output of dada2 into phyloseq object</p>
    </div>

    <pre class="usage"><span class='fu'>import_dada2</span><span class='op'>(</span>
  <span class='va'>seq_tab</span>,
  tax_tab <span class='op'>=</span> <span class='cn'>NULL</span>,
  sam_tab <span class='op'>=</span> <span class='cn'>NULL</span>,
  phy_tree <span class='op'>=</span> <span class='cn'>NULL</span>,
  keep_taxa_rows <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>seq_tab</th>
      <td><p>matrix-like, ASV table, the output of
<code>dada2::removeBimeraDenovo</code>.</p></td>
    </tr>
    <tr>
      <th>tax_tab</th>
      <td><p>matrix, taxonomy table, the output of
<code>dada2::assignTaxonomy</code> or <code>dada2::addSpecies</code>.</p></td>
    </tr>
    <tr>
      <th>sam_tab</th>
      <td><p>data.frame or <code><a href='https://rdrr.io/pkg/phyloseq/man/sample_data-class.html'>phyloseq::sample_data</a></code>, sample data</p></td>
    </tr>
    <tr>
      <th>phy_tree</th>
      <td><p><code><a href='https://rdrr.io/pkg/ape/man/read.tree.html'>ape::phylo</a></code> class or character represents the path of
the tree file</p></td>
    </tr>
    <tr>
      <th>keep_taxa_rows</th>
      <td><p>logical, whether keep taxa in rows or not in the
<code>otu_table</code> of the returned <code>phyloseq</code> object, default <code>TRUE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html'>phyloseq::phyloseq</a></code> object hold the taxonomy info,
sample metadata, number of reads per ASV.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The output of the dada2 pipeline is a feature table of amplicon sequence
variants (an ASV table): A matrix with rows corresponding to samples and
columns to ASVs, in which the value of each entry is the number of times
that ASV was observed in that sample. This table is analogous to the
traditional OTU table. Conveniently, taxa names are saved as ASV1, ASV2,
..., in the returned phyloseq object.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'><span class='va'>seq_tab</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"dada2_seqtab.rds"</span>,</span>
<span class='r-in'>    package <span class='op'>=</span> <span class='st'>"microbiomeMarker"</span></span>
<span class='r-in'><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>tax_tab</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"dada2_taxtab.rds"</span>,</span>
<span class='r-in'>    package <span class='op'>=</span> <span class='st'>"microbiomeMarker"</span></span>
<span class='r-in'><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>sam_tab</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.table</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"dada2_samdata.txt"</span>,</span>
<span class='r-in'>    package <span class='op'>=</span> <span class='st'>"microbiomeMarker"</span></span>
<span class='r-in'><span class='op'>)</span>, sep <span class='op'>=</span> <span class='st'>"\t"</span>, header <span class='op'>=</span> <span class='cn'>TRUE</span>, row.names <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>ps</span> <span class='op'>&lt;-</span> <span class='fu'>import_dada2</span><span class='op'>(</span>seq_tab <span class='op'>=</span> <span class='va'>seq_tab</span>, tax_tab <span class='op'>=</span> <span class='va'>tax_tab</span>, sam_tab <span class='op'>=</span> <span class='va'>sam_tab</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>ps</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> phyloseq-class experiment-level object</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> otu_table()   OTU Table:         [ 232 taxa and 20 samples ]</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> sample_data() Sample Data:       [ 20 samples by 4 sample variables ]</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> tax_table()   Taxonomy Table:    [ 232 taxa by 6 taxonomic ranks ]</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> refseq()      DNAStringSet:      [ 232 reference sequences ]</span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Yang Cao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


