<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ microbiomeMarker</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microbiomeMarker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/microbiomeMarker-vignette.html">Tools for microbiome marker identification</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/yiluheihei/microbiomeMarker/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="data-import">Data import <a href="#data-import" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for importing external data and converting other R object as phyloseq or reverse converting</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="import_dada2.html">import_dada2()</a></code> </p>
        </td>
        <td><p>Import function to read the the output of dada2 as phyloseq object</p></td>
      </tr><tr><td>
          <p><code><a href="import_picrust2.html">import_picrust2()</a></code> </p>
        </td>
        <td><p>Import function to read the output of picrust2 as phyloseq object</p></td>
      </tr><tr><td>
          <p><code><a href="import_qiime2.html">import_qiime2()</a></code> </p>
        </td>
        <td><p>Import function to read the the output of dada2 as phyloseq object</p></td>
      </tr><tr><td>
          <p><code><a href="reexports.html">reexports</a></code> <code><a href="reexports.html">ntaxa</a></code> <code><a href="reexports.html">nsamples</a></code> <code><a href="reexports.html">otu_table</a></code> <code><a href="reexports.html">sample_data</a></code> <code><a href="reexports.html">tax_table</a></code> <code><a href="reexports.html">taxa_names</a></code> <code><a href="reexports.html">sample_names</a></code> <code><a href="reexports.html">%&gt;%</a></code> <code><a href="reexports.html">import_qiime</a></code> <code><a href="reexports.html">import_mothur</a></code> <code><a href="reexports.html">import_biom</a></code> </p>
        </td>
        <td><p>Objects exported from other packages</p></td>
      </tr><tr><td>
          <p><code><a href="phyloseq2DESeq2.html">phyloseq2DESeq2()</a></code> </p>
        </td>
        <td><p>Convert <code>phyloseq-class</code> object to <code>DESeqDataSet-class</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="phyloseq2edgeR.html">phyloseq2edgeR()</a></code> </p>
        </td>
        <td><p>Convert phyloseq data to edgeR <code>DGEList</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="phyloseq2metagenomeSeq.html">phyloseq2metagenomeSeq()</a></code> <code><a href="phyloseq2metagenomeSeq.html">otu_table2metagenomeSeq()</a></code> </p>
        </td>
        <td><p>Convert phyloseq data to MetagenomeSeq <code>MRexperiment</code> object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="microbiome-marker">Microbiome marker <a href="#microbiome-marker" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>S4 class and methods for microbiomeMarker</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="microbiomeMarker.html">microbiomeMarker()</a></code> </p>
        </td>
        <td><p>Build microbiomeMarker-class objects</p></td>
      </tr><tr><td>
          <p><code><a href="microbiomeMarker-class.html">show(<i>&lt;microbiomeMarker&gt;</i>)</a></code> </p>
        </td>
        <td><p>The main class for microbiomeMarker data</p></td>
      </tr><tr><td>
          <p><code><a href="marker_table-class.html">marker_table-class</a></code> </p>
        </td>
        <td><p>The S4 class for storing microbiome marker information</p></td>
      </tr><tr><td>
          <p><code><a href="marker_table-methods.html">marker_table()</a></code> </p>
        </td>
        <td><p>Build or access the marker_table</p></td>
      </tr><tr><td>
          <p><code><a href="assign-marker_table.html">`marker_table&lt;-`()</a></code> </p>
        </td>
        <td><p>Assign marker_table to <code>object</code></p></td>
      </tr><tr><td>
          <p><code><a href="assign-otu_table.html">`otu_table&lt;-`(<i>&lt;microbiomeMarker&gt;</i>,<i>&lt;otu_table&gt;</i>)</a></code> <code><a href="assign-otu_table.html">`otu_table&lt;-`(<i>&lt;microbiomeMarker&gt;</i>,<i>&lt;phyloseq&gt;</i>)</a></code> <code><a href="assign-otu_table.html">`otu_table&lt;-`(<i>&lt;microbiomeMarker&gt;</i>,<i>&lt;microbiomeMarker&gt;</i>)</a></code> </p>
        </td>
        <td><p>Assign a new OTU table</p></td>
      </tr><tr><td>
          <p><code><a href="abundances-methods.html">abundances()</a></code> </p>
        </td>
        <td><p>Extract taxa abundances</p></td>
      </tr><tr><td>
          <p><code><a href="nmarker-methods.html">nmarker()</a></code> </p>
        </td>
        <td><p>Get the number of microbiome markers</p></td>
      </tr><tr><td>
          <p><code><a href="extract-methods.html">`[`(<i>&lt;marker_table&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> </p>
        </td>
        <td><p>Extract <code>marker_table</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="postHocTest-class.html">show(<i>&lt;postHocTest&gt;</i>)</a></code> </p>
        </td>
        <td><p>The postHocTest Class, represents the result of post-hoc test result among
multiple groups</p></td>
      </tr><tr><td>
          <p><code><a href="postHocTest.html">postHocTest()</a></code> </p>
        </td>
        <td><p>Build postHocTest object</p></td>
      </tr><tr><td>
          <p><code><a href="extract_posthoc_res.html">extract_posthoc_res()</a></code> </p>
        </td>
        <td><p>Extract results from a posthoc test</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="functions-reexports-from-phyloseq">Functions reexports from phyloseq <a href="#functions-reexports-from-phyloseq" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="reexports.html">reexports</a></code> <code><a href="reexports.html">ntaxa</a></code> <code><a href="reexports.html">nsamples</a></code> <code><a href="reexports.html">otu_table</a></code> <code><a href="reexports.html">sample_data</a></code> <code><a href="reexports.html">tax_table</a></code> <code><a href="reexports.html">taxa_names</a></code> <code><a href="reexports.html">sample_names</a></code> <code><a href="reexports.html">%&gt;%</a></code> <code><a href="reexports.html">import_qiime</a></code> <code><a href="reexports.html">import_mothur</a></code> <code><a href="reexports.html">import_biom</a></code> </p>
        </td>
        <td><p>Objects exported from other packages</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="normalization">Normalization <a href="#normalization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="abundances-methods.html">abundances()</a></code> </p>
        </td>
        <td><p>Extract taxa abundances</p></td>
      </tr><tr><td>
          <p><code><a href="transform_abundances.html">transform_abundances()</a></code> </p>
        </td>
        <td><p>Transform the taxa abundances in <code>otu_table</code> sample by sample</p></td>
      </tr><tr><td>
          <p><code><a href="normalize-methods.html">normalize(<i>&lt;phyloseq&gt;</i>)</a></code> <code><a href="normalize-methods.html">normalize(<i>&lt;otu_table&gt;</i>)</a></code> <code><a href="normalize-methods.html">normalize(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="normalize-methods.html">normalize(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="normalize-methods.html">norm_rarefy()</a></code> <code><a href="normalize-methods.html">norm_tss()</a></code> <code><a href="normalize-methods.html">norm_css()</a></code> <code><a href="normalize-methods.html">norm_rle()</a></code> <code><a href="normalize-methods.html">norm_tmm()</a></code> <code><a href="normalize-methods.html">norm_clr()</a></code> <code><a href="normalize-methods.html">norm_cpm()</a></code> </p>
        </td>
        <td><p>Normalize the microbial abundance data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="differential-analysis">Differential analysis <a href="#differential-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for identifying the microbiome markers</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="confounder.html">confounder()</a></code> </p>
        </td>
        <td><p>Confounder analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_aldex.html">run_aldex()</a></code> </p>
        </td>
        <td><p>Perform differential analysis using ALDEx2</p></td>
      </tr><tr><td>
          <p><code><a href="run_ancom.html">run_ancom()</a></code> </p>
        </td>
        <td><p>Perform differential analysis using ANCOM</p></td>
      </tr><tr><td>
          <p><code><a href="run_ancombc.html">run_ancombc()</a></code> </p>
        </td>
        <td><p>Differential analysis of compositions of microbiomes with bias correction
(ANCOM-BC).</p></td>
      </tr><tr><td>
          <p><code><a href="run_deseq2.html">run_deseq2()</a></code> </p>
        </td>
        <td><p>Perform DESeq differential analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_edger.html">run_edger()</a></code> </p>
        </td>
        <td><p>Perform differential analysis using edgeR</p></td>
      </tr><tr><td>
          <p><code><a href="run_lefse.html">run_lefse()</a></code> </p>
        </td>
        <td><p>Liner discriminant analysis (LDA) effect size (LEFSe) analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_limma_voom.html">run_limma_voom()</a></code> </p>
        </td>
        <td><p>Differential analysis using limma-voom</p></td>
      </tr><tr><td>
          <p><code><a href="run_marker.html">run_marker()</a></code> </p>
        </td>
        <td><p>Find makers (differentially expressed metagenomic features)</p></td>
      </tr><tr><td>
          <p><code><a href="run_metagenomeseq.html">run_metagenomeseq()</a></code> </p>
        </td>
        <td><p>metagenomeSeq differential analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_posthoc_test.html">run_posthoc_test()</a></code> </p>
        </td>
        <td><p>Post hoc pairwise comparisons for multiple groups test.</p></td>
      </tr><tr><td>
          <p><code><a href="run_simple_stat.html">run_simple_stat()</a></code> </p>
        </td>
        <td><p>Simple statistical analysis of metagenomic profiles</p></td>
      </tr><tr><td>
          <p><code><a href="run_sl.html">run_sl()</a></code> </p>
        </td>
        <td><p>Identify biomarkers using supervised leaning (SL) methods</p></td>
      </tr><tr><td>
          <p><code><a href="run_test_multiple_groups.html">run_test_multiple_groups()</a></code> </p>
        </td>
        <td><p>Statistical test for multiple groups</p></td>
      </tr><tr><td>
          <p><code><a href="run_test_two_groups.html">run_test_two_groups()</a></code> </p>
        </td>
        <td><p>Statistical test between two groups</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="comparison-of-different-methods">Comparison of different methods <a href="#comparison-of-different-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="compare_DA.html">compare_DA()</a></code> </p>
        </td>
        <td><p>Comparing the results of differential analysis methods by Empirical power
and False Discovery Rate</p></td>
      </tr><tr><td>
          <p><code><a href="summary.compareDA.html">summary(<i>&lt;compareDA&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summary differential analysis methods comparison results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="effect_size-plot.html">plot_ef_bar()</a></code> <code><a href="effect_size-plot.html">plot_ef_dot()</a></code> </p>
        </td>
        <td><p>bar and dot plot of effect size of microbiomeMarker data</p></td>
      </tr><tr><td>
          <p><code><a href="plot.compareDA.html">plot(<i>&lt;compareDA&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plotting DA comparing result</p></td>
      </tr><tr><td>
          <p><code><a href="plot_abundance.html">plot_abundance()</a></code> </p>
        </td>
        <td><p>plot the abundances of markers</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cladogram.html">plot_cladogram()</a></code> </p>
        </td>
        <td><p>plot cladogram of micobiomeMaker results</p></td>
      </tr><tr><td>
          <p><code><a href="plot_heatmap.html">plot_heatmap()</a></code> </p>
        </td>
        <td><p>Heatmap of microbiome marker</p></td>
      </tr><tr><td>
          <p><code><a href="plot_postHocTest.html">plot_postHocTest()</a></code> </p>
        </td>
        <td><p><code>postHocTest</code> plot</p></td>
      </tr><tr><td>
          <p><code><a href="plot_sl_roc.html">plot_sl_roc()</a></code> </p>
        </td>
        <td><p>ROC curve of microbiome marker from supervised learning methods</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="example-data">Example data <a href="#example-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="data-caporaso.html">data-caporaso</a></code> <code><a href="data-caporaso.html">caporaso</a></code> </p>
        </td>
        <td><p>16S rRNA data from "Moving pictures of the human microbiome"</p></td>
      </tr><tr><td>
          <p><code><a href="data-cid_ying.html">data-cid_ying</a></code> <code><a href="data-cid_ying.html">cid_ying</a></code> </p>
        </td>
        <td><p>16S rRNA data of 94 patients from CID 2012</p></td>
      </tr><tr><td>
          <p><code><a href="data-ecam.html">data-ecam</a></code> <code><a href="data-ecam.html">ecam</a></code> </p>
        </td>
        <td><p>Data from Early Childhood Antibiotics and the Microbiome (ECAM) study</p></td>
      </tr><tr><td>
          <p><code><a href="data-enterotypes_arumugam.html">data-enterotypes_arumugam</a></code> <code><a href="data-enterotypes_arumugam.html">enterotypes_arumugam</a></code> </p>
        </td>
        <td><p>Enterotypes data of 39 samples</p></td>
      </tr><tr><td>
          <p><code><a href="data-kostic_crc.html">data-kostic_crc</a></code> <code><a href="data-kostic_crc.html">kostic_crc</a></code> </p>
        </td>
        <td><p>Data from a study on colorectal cancer (kostic 2012)</p></td>
      </tr><tr><td>
          <p><code><a href="data-oxygen.html">data-oxygen</a></code> <code><a href="data-oxygen.html">oxygen</a></code> </p>
        </td>
        <td><p>Oxygen availability 16S dataset, of which taxa table has been summarized for
python lefse input</p></td>
      </tr><tr><td>
          <p><code><a href="data-pediatric_ibd.html">data-pediatric_ibd</a></code> <code><a href="data-pediatric_ibd.html">pediatric_ibd</a></code> </p>
        </td>
        <td><p>IBD stool samples</p></td>
      </tr><tr><td>
          <p><code><a href="data-spontaneous_colitis.html">data-spontaneous_colitis</a></code> <code><a href="data-spontaneous_colitis.html">spontaneous_colitis</a></code> </p>
        </td>
        <td><p>This is a sample data from lefse python script, a 16S dataset for studying
the characteristics of the fecal microbiota in a mouse model of spontaneous
colitis.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="miscellaneous">Miscellaneous <a href="#miscellaneous" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="aggregate_taxa.html">aggregate_taxa()</a></code> </p>
        </td>
        <td><p>Aggregate Taxa</p></td>
      </tr><tr><td>
          <p><code><a href="subset_marker.html">subset_marker()</a></code> </p>
        </td>
        <td><p>Subset microbiome markers</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_taxa.html">summarize_taxa()</a></code> </p>
        </td>
        <td><p>Summarize taxa into a taxonomic level within each sample</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yang Cao.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

