<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform differential analysis using ANCOM — run_ancom • microbiomeMarker</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Perform differential analysis using ANCOM — run_ancom" />
<meta property="og:description" content="Perform significant test by comparing the pairwise log ratios between all
features." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microbiomeMarker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/microbiomeMarker-vignette.html">Tools for microbiome marker identification</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/yiluheihei/microbiomeMarker/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform differential analysis using ANCOM</h1>
    <small class="dont-index">Source: <a href='https://github.com/yiluheihei/microbiomeMarker/blob/master/R/DA-ancom.R'><code>R/DA-ancom.R</code></a></small>
    <div class="hidden name"><code>run_ancom.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform significant test by comparing the pairwise log ratios between all
features.</p>
    </div>

    <pre class="usage"><span class='fu'>run_ancom</span><span class='op'>(</span>
  <span class='va'>ps</span>,
  <span class='va'>group</span>,
  taxa_rank <span class='op'>=</span> <span class='st'>"all"</span>,
  transform <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"identity"</span>, <span class='st'>"log10"</span>, <span class='st'>"log10p"</span><span class='op'>)</span>,
  norm <span class='op'>=</span> <span class='st'>"TSS"</span>,
  norm_para <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='op'>)</span>,
  p_adjust <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"none"</span>, <span class='st'>"fdr"</span>, <span class='st'>"bonferroni"</span>, <span class='st'>"holm"</span>, <span class='st'>"hochberg"</span>, <span class='st'>"hommel"</span>, <span class='st'>"BH"</span>, <span class='st'>"BY"</span><span class='op'>)</span>,
  pvalue_cutoff <span class='op'>=</span> <span class='fl'>0.05</span>,
  W_cutoff <span class='op'>=</span> <span class='fl'>0.75</span>,
  test <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"aov"</span>, <span class='st'>"wilcox.test"</span>, <span class='st'>"kruskal.test"</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ps</th>
      <td><p>a <code><a href='https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html'>phyloseq-class</a></code> object.</p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>character, the variable to set the group.</p></td>
    </tr>
    <tr>
      <th>taxa_rank</th>
      <td><p>character to specify taxonomic rank to perform
differential analysis on. Should be one of
<code><a href='https://rdrr.io/pkg/phyloseq/man/rank_names.html'>phyloseq::rank_names(phyloseq)</a></code>, or "all" means to summarize the taxa by
the top taxa ranks (<code><a href='summarize_taxa.html'>summarize_taxa(ps, level = rank_names(ps)[1])</a></code>), or
"none" means perform differential analysis on the original taxa
(<code><a href='https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html'>taxa_names(phyloseq)</a></code>, e.g., OTU or ASV).</p></td>
    </tr>
    <tr>
      <th>transform</th>
      <td><p>character, the methods used to transform the microbial
abundance. See <code><a href='transform_abundances.html'>transform_abundances()</a></code> for more details. The
options include:</p><ul>
<li><p>"identity", return the original data without any transformation.</p></li>
<li><p>"log10", the transformation is <code><a href='https://rdrr.io/r/base/Log.html'>log10(object)</a></code>, and if the data contains
zeros the transformation is <code><a href='https://rdrr.io/r/base/Log.html'>log10(1 + object)</a></code>.</p></li>
<li><p>"log10p", the transformation is <code><a href='https://rdrr.io/r/base/Log.html'>log10(1 + object)</a></code>.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>norm</th>
      <td><p>the methods used to normalize the microbial abundance data. See
<code><a href='normalize,phyloseq-method.html'>normalize()</a></code> for more details.
Options include:</p><ul>
<li><p>"none": do not normalize.</p></li>
<li><p>"rarefy": random subsampling counts to the smallest library size in the
data set.</p></li>
<li><p>"TSS": total sum scaling, also referred to as "relative abundance", the
abundances were normalized by dividing the corresponding sample library
size.</p></li>
<li><p>"TMM": trimmed mean of m-values. First, a sample is chosen as reference.
The scaling factor is then derived using a weighted trimmed mean over
the differences of the log-transformed gene-count fold-change between
the sample and the reference.</p></li>
<li><p>"RLE", relative log expression, RLE uses a pseudo-reference calculated
using the geometric mean of the gene-specific abundances over all
samples. The scaling factors are then calculated as the median of the
gene counts ratios between the samples and the reference.</p></li>
<li><p>"CSS": cumulative sum scaling, calculates scaling factors as the
cumulative sum of gene abundances up to a data-derived threshold.</p></li>
<li><p>"CLR": centered log-ratio normalization.</p></li>
<li><p>"CPM": pre-sample normalization of the sum of the values to 1e+06.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>norm_para</th>
      <td><p>named <code>list</code>. other arguments passed to specific
normalization methods.  Most users will not need to pass any additional
arguments here.</p></td>
    </tr>
    <tr>
      <th>p_adjust</th>
      <td><p>method for multiple test correction, default <code>none</code>,
for more details see <a href='https://rdrr.io/r/stats/p.adjust.html'>stats::p.adjust</a>.</p></td>
    </tr>
    <tr>
      <th>pvalue_cutoff</th>
      <td><p>significance level for each of the statistical tests,
default 0.05.</p></td>
    </tr>
    <tr>
      <th>W_cutoff</th>
      <td><p>lower bound for the proportion for the W-statistic, default
0.7.</p></td>
    </tr>
    <tr>
      <th>test</th>
      <td><p>character, the test to dtermine the p value of log ratio,
one of "aov", "wilcox.test",  "kruskal.test".</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments passed to the test function.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <a href='microbiomeMarker-class.html'>microbiomeMarker</a> object, in which the <code>slot</code> of
<code>marker_table</code> contains four variables:</p><ul>
<li><p><code>feature</code>, significantly different features.</p></li>
<li><p><code>enrich_group</code>, the class of the differential features enriched.</p></li>
<li><p><code>effect_size</code>, differential means for two groups, or F statistic for more
than two groups.</p></li>
<li><p><code>W</code>, the W-statistic, number of features that a single feature is tested
to be significantly different against.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>In an experiment with only two treatments, this tests the following
hypothesis for feature \(i\):</p>
<p>$$H_{0i}: E(log(\mu_i^1)) =  E(log(\mu_i^2))$$</p>
<p>where \(\mu_i^1\) and \(\mu_i^2\) are the mean abundances for feature
\(i\) in the two groups.</p>
<p>The developers of this method recommend the following significance tests
if there are 2 groups, use non-parametric Wilcoxon rank sum test
<code><a href='https://rdrr.io/r/stats/wilcox.test.html'>stats::wilcox.test()</a></code>. If there are more than 2 groups, use nonparametric
<code><a href='https://rdrr.io/r/stats/kruskal.test.html'>stats::kruskal.test()</a></code> or one-way ANOVA <code><a href='https://rdrr.io/r/stats/aov.html'>stats::aov()</a></code>.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Mandal et al. "Analysis of composition of microbiomes: a novel
method for studying microbial composition", Microbial Ecology in Health
&amp; Disease, (2015), 26.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Huang Lin, Yang Cao</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># \donttest{</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>enterotypes_arumugam</span><span class='op'>)</span>
<span class='va'>ps</span> <span class='op'>&lt;-</span> <span class='fu'>phyloseq</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html'>subset_samples</a></span><span class='op'>(</span>
  <span class='va'>enterotypes_arumugam</span>,
  <span class='va'>Enterotype</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Enterotype 3"</span>, <span class='st'>"Enterotype 2"</span><span class='op'>)</span>
<span class='op'>)</span>
<span class='fu'>run_ancom</span><span class='op'>(</span><span class='va'>ps</span>, group <span class='op'>=</span> <span class='st'>"Enterotype"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; microbiomeMarker-class inherited from phyloseq-class
#&gt; normalization method:              [ TSS ]
#&gt; microbiome marker identity method: [ ANCOM ]
#&gt; marker_table() Marker Table:       [ 8 microbiome markers with 4 variables ]
#&gt; otu_table()    OTU Table:          [ 235 taxa and  24 samples ]
#&gt; sample_data()  Sample Data:        [ 24 samples by  9 sample variables ]
#&gt; tax_table()    Taxonomy Table:     [ 235 taxa by 1 taxonomic ranks ]</div><div class='input'><span class='co'># }</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Yang Cao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


